---
title: "Mariner: Explore the Hi-Cs"
author: "Eric S. Davis, Sarah M. Parker, Manjari Kiran, \
Nicole Kramer, Douglas H. Phanstiel"
date: '`r Sys.Date()`'
output: rmarkdown::html_vignette
bibliography: references.bib
link-citations: true
vignette: >
  %\VignetteIndexEntry{Mariner: Explore the Hi-Cs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse=TRUE,
  comment="#>"
)
```

```{r, include=FALSE}
## List the contents & files of the s3 data
contents <- aws.s3::get_bucket("s3://bioc2023mariner", region="us-east-2")
files <- vapply(contents, "[[", j="Key", FUN.VALUE=character(1L)) |> unname()

## Write a function to download required data from s3
download_file <- function(file) {
  
  ## If object doesn't exist in bucket, return NA
  ok <- aws.s3::object_exists(
    object=file,
    bucket="s3://bioc2023mariner/", 
    region="us-east-2"
  )
  if(!ok) return(NA_character_)
  
  ## If object exists, save it and return file path
  aws.s3::save_object(
      object=file,
      bucket="s3://bioc2023mariner/", 
      region="us-east-2",
      file=file
  )
}

## Download files if they don't already exist
for(f in files) {
  if (!file.exists(f)) {
    download_file(f)
  }
}
```


## Overview

### Description

3D chromatin structure plays an integral, yet incompletely
understood role in the long-distance regulation of genes by
enhancers and repressors. Disruption or aberrant formation of
these long-range interactions can result in developmental
abnormalities and diseases, such as cancer. Therefore, deriving
biological insights from 3D chromatin structure experiments,
such as Hi-C or Micro-C, is essential for understanding and
correcting human disease.

In this workshop, we will use a recently published dataset 
[@Bond2022.10.31.514600] exploring chromatin loop dynamics
during a megakaryocyte differentiation event to demonstrate a
differential looping analysis with `mariner`, `DESeq2`, and
`plotgardener`. Specifically, we will show how `mariner` can
be used to merge redundant loop calls and extract Hi-C contact
frequency between putative loop anchors to use with `DESeq2` to
identify differential loops. We will then conduct pileup analyses
of Hi-C pixels, domains, and boundaries with `mariner`. Finally,
we will demonstrate how these analyses can be visualized together
in a publication-ready multi-panel figure with `plotgardener`.

### Pre-requisites

* Basic knowledge of R syntax
* Familiarity with the `GenomicRanges` and `InteractionSet` classes

### Optional background reading

* Biological background
  * Hi-C background [@rao_3d_2014]
  * Dataset used in workshop [@Bond2022.10.31.514600]
* Other _Bioconductor_ packages
  * `InteractionSet` [@InteractionSet]
  * `DelayedArray` [@DelayedArray]
  * `DESeq2` [@DESeq2; @apeglm]
  * `plotgardener` [@plotgardener]

### Participation

The format is lecture + lab, with time for data exploration.

### Major _R_ / _Bioconductor_ packages used

* `mariner`
* `DESeq2`
* `plotgardener`

### Time outline

90-minute workshop:

| Activity                            | Time |
|-------------------------------------|------|
| Overview of Hi-C data               | 10m  |
| Workshop datasets                   |  5m  |
| Differential analysis with `DESeq2` | 25m  |
| Aggregating interactions            | 25m  |
| Visualization with `plotgardener`   | 25m  |


### Workshop goals and objectives

### Learning goals

* Understand how to work with interaction data in R.
* Learn which `mariner` functions to use for extracting
Hi-C data.
* How to use accessors to get information from `mariner`
objects.
* Integrating `mariner` with `plotgardener` for 
programmatically visualizing genomic data.

### Learning objectives

* Conduct differential interaction analysis with `mariner` and `DESeq2`.
* Make aggregate plots with `mariner`.
* Visualize aggregate plots and genomic regions with `plotgardener`.

## Overview of Hi-C data

<iframe src="https://ericscottdavis.com/HowToHic" style="border: none; width: 100%; height: 500px;"></iframe>

### Terminology 

* *anchors*: the genomic loci (genomic ranges) that make
up an _interaction_.

* *interactions*: a pair of genomic loci (genomic ranges).

* *pixel*: a two-dimensional bin that represents
the _interaction_ between two fixed-width genomic loci.

* *contact frequency*: the number of Hi-C reads assigned
to a pixel.

* *loop*: a _pixel_ that is enriched for _contact frequency_
compared to its local background. Biologically, a _loop_
represents a point of frequent or sustained contact between
genomic loci.

* *the diagonal*: points of self-interaction along the Hi-C
matrix.

* *upper triangular*: the portion of the Hi-C matrix above 
_the diagonal_. _Interactions_ where the first _anchor_
is less than the second _anchor_ appear here.

* *lower triangular*: the portion of the Hi-C matrix below
_the diagonal_. _Interactions_ where the first _anchor_
is greater than the second _anchor_ appear here.

* *resolution*: the ability to distinguish between two
neighboring points on a Hi-C matrix. The higher the
_resolution_ the more precise the contact between
pairs of loci.

* *binSize*: Hi-C reads are quantified along the genome
in regular bins. The size of these bins determines the
_resolution_ of the Hi-C matrix. Larger _bins_ lead
to lower _resolution_ while smaller _bins_ mean higher
_resolution_. Since the quantification of reads into
bins is time consuming, `.hic` files are processed at
several pre-determined `binSizes` (also called 
_resolutions_).

## Workshop datasets

The dataset used in this workshop comes from one of our
recently accepted papers, led by the talented Marielle
Bond, where we explore chromatin loop dynamics during
megakaryocyte differentiation [@Bond2022.10.31.514600].
In short, deeply sequenced Hi-C was conducted in K562 cells
before and after treatment with PMA for 0, 6, or 72 hours.
RNA-seq, ATAC-seq, and CUT&RUN for H3K27ac was also performed
at each timepoint, but will not be used in this workshop.

```{r, fig.align='center', out.width='75%', echo=FALSE}
knitr::include_graphics("img/mega_slide.png")
```

The Hi-C data in this study was processed using a modified
version of the `Juicer` pipeline [@durand_juicer_2016; 
@eric_davis_2022_7514346] resulting in `.hic` files for each
biological and technical replicate. Loops were identified
using [`SIP`](https://github.com/PouletAxel/SIP) [@Rowley01032020]
for each timepoint.

Since the full dataset is very large, we will be working with
a subset of these data in the workshop. We will use 6 Hi-C
replicates (ignoring biological/technical designations) for the
control (0 hours of PMA) and 6 replicates for the treatment
(72 hours of PMA). The `.hic` files and loop files have been
filtered to only contain reads from the first chromosome
of the human hg38 genome build.

The following shows the directory structure of the files
used in this workshop:

```bash
vignettes/data/
├── hic
│   ├── "0hr_1_1_chr1.hic"
│   ├── "0hr_1_2_chr1.hic"
│   ├── "0hr_2_1_chr1.hic"
│   ├── "0hr_2_2_chr1.hic"
│   ├── "0hr_3_1_chr1.hic"
│   ├── "0hr_3_2_chr1.hic"
│   ├── "72hr_1_1_chr1.hic"
│   ├── "72hr_1_2_chr1.hic"
│   ├── "72hr_2_1_chr1.hic"
│   ├── "72hr_2_2_chr1.hic"
│   ├── "72hr_3_1_chr1.hic"
│   └── "72hr_3_2_chr1.hic"
├── loops
│   ├── "0hr_5kbLoops_chr1.txt"
│   ├── "6hr_5kbLoops_chr1.txt"
│   └── "72hr_5kbLoops_chr1.txt"
└── samples
    └── "samplesheet.txt"
```

## Differential analysis with `DESeq2`

In this section we will use `DESeq2` and `mariner` to
identify differential loops between K562 cells treated
with PMA for 0 or 72 hours.

### Merge redundant loops

The first step in this analysis is to read in and combine
loop calls for K562 cells treated for 0, 6, or 72 hours.
Due to technical and biological variability, loop callers
can identify slightly different pixels as loops in different
datasets. Sometimes this represents a true biological shift
in the position of the loop anchors, but more often is due to
noise and technical variability. Furthermore, testing these
redundant interactions reduces our power to identify true 
differences in looping.

```{r mergingFig, fig.align='center', out.width='75%', echo=FALSE}
knitr::include_graphics("img/cluster_merge.png")
```

To combat this effect, we can use `mariner` functions to 
cluster then combine loops between conditions. First, we
start by reading in BEDPE-formatted loops, and converting
them to `GInteractions` objects.

```{r message=FALSE}
library(mariner)

## Read in loops
loops00hr <- read.table("data/loops/0hr_5kbLoops_chr1.txt", header=TRUE)
loops06hr <- read.table("data/loops/6hr_5kbLoops_chr1.txt", header=TRUE)
loops72hr <- read.table("data/loops/72hr_5kbLoops_chr1.txt", header=TRUE)

## Put into a list
loopList <- list("0hr"=loops00hr, "6hr"=loops06hr, "72hr"=loops72hr)

## Convert to GInteractions objects
loopList <- lapply(loopList, as_ginteractions)

## Show list of GInteractions
lapply(loopList, summary)
```
```{r}
loopList[["0hr"]]
```

Then we can use the `mergePairs()` function to cluster
then merge the list of `GInteraction` objects.

This function uses `dbscan` to cluster loops by a given
genomic `radius` in base-pairs. It selects a representative
interaction for each cluster using either the metadata
`column` provided, or if `column=NULL` it calculates new
ranges for each pair by calculating the median of modes.

```{r mergingFig2, fig.align='center', out.width='50%', echo=FALSE}
knitr::include_graphics("img/cluster_merge2.png")
```

```{r}
mergedLoops <- mergePairs(x=loopList, radius=10e3, column="APScoreAvg")
mergedLoops
```

### Exercise - Finding de novo loops

Let's take a closer look at the `MergedGInteractions` object.
This object looks and behaves as a normal `GInteractions`
object, but also contains slots storing the clustered
interactions and associated metadata. There are a few key
accessor functions that allow you to extract, aggregate, and
subset the interactions according to how they have clustered.

The input to `mergePairs()` was a list of `GInteractions` objects.
The `sources()` accessor returns the names of each input source
file:

```{r}
sources(mergedLoops)
```

Each row of the `MergedGInteractions` object represents
the selected interaction from each cluster of one or more
interactions. To view all interactions in a cluster, you
can use the `clusters()` function:

```{r}
clusters(mergedLoops[1500])
```

Notice that it returns a `data.table` with a column appended to the
end, `src`, indicating the source file of each interaction. All of the
metadata for each specific interaction is also included. If you want
to summarize one or more of these values for each cluster in the object,
the `aggMetadata()` function is handy.

```{r}
aggMetadata(x=mergedLoops, columns="APScoreAvg", funs="mean")
```

Finally, we can subset these interactions by the source file
using the `sets()` function. This function includes
optional parameters, `include` and `exclude`, which control which
sources should be grouped in the output. By default a list of
`MergedGInteractions` objects is returned that contains the set
of interactions belonging to each combination of sources.

```{r}
sets(x=mergedLoops) |>
  lapply(summary)
```

We can visualize these sets with an `UpSet` plot:

```{r}
library(UpSetR)

## Count loops in each set
loopSets <- 
  sets(x=mergedLoops) |>
  lapply(length)

## Convert to an expression set for UpSetR
loopSets <-
  setNames(loopSets, gsub("_", "&", names(loopSets))) |>
  fromExpression()

## Visualize
upset(loopSets, order.by="freq")
```

If we want to find loops that form _de_ _novo_ after 72 hours
explicitly excluding the 0 and 6 hour time points, we can use the
following code:

```{r}
sets(x=mergedLoops, include="72hr", exclude=c("0hr", "6hr"))
```
To find loops that were present at either 6 or 72 hrs, but not
present at 0 hrs, we can exclude the 0hr time point:

```{r}
sets(x=mergedLoops, exclude="0hr")
```

*Exercise*: How could we change this code to find
transient loops, or loops that appeared after 6 hours of PMA
treatment and dissolved after 72 hours of treatment?

```{r, results='hide', eval=FALSE, echo=FALSE}
sets(x=mergedLoops, include="6hr", exclude=c("0hr", "72hr"))
```

Now let's return to our differential looping analysis to
learn how to find quantitative changes in looping.

### Finding differential loops

In the previous section, we found loops that formed
_de_ _novo_ after treatment with PMA for 0, 6, or 72
hours. Using this approach relies entirely on the loop
caller's threshold for calling loops, but is unable to
find quantitative changes - such as the strengthening
or weakening of existing loops between conditions. 
Additionally, the presence/absence of a loop doesn't 
necessarily mean that we gained or lost a loop. In this
section we take a more statistically rigorous approach to
finding differential looping.

To do this we will use `DESeq2` to conduct the
differential analysis. While `DEseq2` is typically
used for RNA-seq analysis, we find that it works very well
for finding differential Hi-C interactions. `DEseq2` requires
three pieces of information to run: 1) a count matrix where
each row represents a loop and each column represents a Hi-C
replicate sample, 2) a `data.frame` called `colData` that
contains sample-specific information, and 3) a design formula
that describes the model to test.

We can use `mariner` to create the count matrix by extracting
unnormlized contact frequency between interactions for each
Hi-C sample. First we read in the sample sheet containing
Hi-C sample information and file paths each `.hic` file:
  
```{r}
samples <- read.table("data/samples/samplesheet.txt", header=TRUE, colClasses="factor")
samples <- as(samples, "DataFrame")
samples
```

Since we want a count matrix with a single value for
each 5 Kb loop pixel and `.hic` file, the `pullHicPixels()`
function is what we should use with a `binSize` of 5 Kb:
  
```{r}
## Hi-C file paths from samplesheet
hicFiles <- as.character(samples$filepath)

## Extract pixels
pixels <- pullHicPixels(
  x=mergedLoops,
  files=hicFiles,
  binSize=5e3,
  blockSize=100e6
)
pixels
```

The object that results is called an `InteractionMatrix`
because it returns a single matrix for each interaction
and `.hic` file. You can access this matrix with the
`counts()` accessor:
  
```{r}
counts(pixels)
```

```{r interactionMatrixFig, fig.align='center', out.width='50%', echo=FALSE}
knitr::include_graphics("img/interactionMatrixFig.png")
```

These counts are stored on-disk in an HDF5 file to
conserve R-memory usage. But we will talk more about
this later. For now, we can create a `DESeqDataSet` and
run `DESeq()`:
  
```{r, message=FALSE}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(
  countData=counts(pixels),
  colData=cbind(colData(pixels), samples),
  design=~replicate+condition
)
dds <- DESeq(dds)
dds
```

Get shrunken results:
  
```{r, message=FALSE}
res <- lfcShrink(dds, coef="condition_treatment_vs_control", type="apeglm")
summary(res)
```
And inspect the results with an MA plot:
  
```{r}
plotMA(res)
```

or a PCA plot:
  
```{r}
varianceStabilizingTransformation(dds) |>
  plotPCA(intgroup="condition") +
  ggplot2::theme(aspect.ratio = 1)
```

We can add these results from `DESeq2` to our
`InteractionMatrix` object.

```{r}
rowData(pixels) <- res
pixels
```

```{r, results='hide', eval=FALSE, echo=FALSE}
## Example of using DESeqDataSet contructor
tmp <- pixels
library(SummarizedExperiment)
colData(tmp) <- cbind(colData(tmp), samples)
counts(tmp) <- as.matrix(counts(tmp))

library(DESeq2)
DESeqDataSet(tmp, design = ~replicate+condition)
```

## Aggregating interactions

`mariner` is not limited to extracting single pixels.
The `pullHicMatrices()` function allows you to
define square or rectangular regions and extract them
as matrices divided into specified bins. For example,
the figure below shows that supplying a `binSize` that
is half of each anchor width produces a 2x2 matrix of
pixels.

```{r matricesFigure, fig.align='center', out.width='50%', echo=FALSE}
knitr::include_graphics("img/matricesFigure.png")
```

In the following sections, we will define square regions
around our merged loop pixels and extract count matrices.
We will then aggregate these matrices and visualize the
results in what is called an Aggregate Peak Analysis (APA).
APA plots are essentially pile-up signal plots for Hi-C.
APA is helpful for identifying genome-wide trends at specific
loci, such as loop anchors or protein binding sites.

### Aggregate peak analysis of differential loops

We can use a helper function called `pixelsToMatrices()`
to expand square regions around each central pixel.

```{r pixToMat, fig.align='center', out.width='10%', echo=FALSE}
knitr::include_graphics("img/pixToMat.png")
```

```{r}
regions <- pixelsToMatrices(x=interactions(pixels), buffer=5)
regions
```

Then we can use the `pullHicMatrices()` function to extract
11 x 11 matrices by setting the `binSize` to 5 Kb:
  
```{r}
matrices <- pullHicMatrices(
  x=regions,
  files=hicFiles,
  binSize=5e3,
  blockSize=100e6
)
matrices
```

This results in an `InteractionArray` object that is very
similar to the `InteractionMatrix` object except that it
contains an array of pixel by Hi-C count matrices that
are accessible with the `counts()` accessor:
  
```{r}
counts(matrices)
```

This array contains count matrices in the first two
dimensions, interactions in the third dimension,
and Hi-C files in the fourth dimension. The data is
stored in an HDF5 file that is accessible with the
`DelayedArray` package. This allows us to use this
object as if it were held in R's memory, but only
loading data as needed (i.e. when operations are
applied to it).

```{r interactionArrayFig, fig.align='center', out.width='75%', echo=FALSE}
knitr::include_graphics("img/interactionArrayFig.png")
```


We can then aggregate this array with the fast
and efficient `aggHicMatrices()` function.
This allows you to aggregate by interactions,
files, or all.

```{r aggMatFigAlt, fig.align='center', out.width='75%', echo=FALSE}
knitr::include_graphics("img/aggMatFigAlt.png")
```

Here we will combine all interactions for each
Hi-C file by setting `by="files"`:

```{r}
aggMats <- aggHicMatrices(x=matrices, by="files")
aggMats
```

*Exercise* - Try different values for the `by` argument
and see what you get!

Since the result is a `DelayedArray` object, we can
use `apply()` functions to do more complex aggregations.
Let's combine all the biological and technical replicates
for each condition (0hr or 72hr).

```{r}
ctr <- apply(aggMats[,,1:6], c(1,2), sum, na.rm=TRUE)
trt <- apply(aggMats[,,7:12], c(1,2), sum, na.rm=TRUE)
```

And visualize them with the `plotMatrix()` function:
  
```{r, message=FALSE, fig.show="hold", out.width="50%"}
plotMatrix(ctr)
plotMatrix(trt)
```

You'll notice that there is an artifact in the bottom
left-hand corner. This is because interactions that are
short, or close to the diagonal, are being superimposed.
The `removeShortPairs()` function can be used upstream
of `pullHicMatrices()` function to remove any part of the regions
that would cross the diagonal. The following code shows how
these functions can be chained to produce APA plots without
artifacts caused by the Hi-C diagonal.

```{r, message=FALSE, fig.show="hold", out.width="50%"}
## Remove short pairs & extract counts
matrices <-
  regions |>
  removeShortPairs() |>
  pullHicMatrices(files=hicFiles, binSize=5e3, blockSize=100e6)

## Aggregate by files
aggMats <- aggHicMatrices(matrices, by="files")

## Combine bio & tech replicates
ctr <- apply(aggMats[,,1:6], c(1,2), sum, na.rm=TRUE)
trt <- apply(aggMats[,,7:12], c(1,2), sum, na.rm=TRUE)

## Visualize without artifacts
plotMatrix(ctr)
plotMatrix(trt)
```

You may notice that there is not much difference
between the APA of the control vs. treatment because
we are aggregating all loops together. We can subset
the `InteractionArray` object according to our
differential looping results to show genome-wide
differences that support our analysis.

```{r, message=FALSE}
## Subset into gained/lost loops
res <- rowData(matrices)
gained <- matrices[which(res$padj <= 0.1 & res$log2FoldChange > 0)]
lost <- matrices[which(res$padj <= 0.1 & res$log2FoldChange < 0)]

## Aggregate gained/lost for control & treated
gained_ctr <- apply(counts(gained)[,,,1:6], c(1,2), sum, na.rm=TRUE)
gained_trt <- apply(counts(gained)[,,,7:12], c(1,2), sum, na.rm=TRUE)
lost_ctr <- apply(counts(lost)[,,,1:6], c(1,2), sum, na.rm=TRUE)
lost_trt <- apply(counts(lost)[,,,7:12], c(1,2), sum, na.rm=TRUE)

## Normalize to the number of interactions
gained_ctr <- gained_ctr/length(gained)
gained_trt <- gained_trt/length(gained)
lost_ctr <- lost_ctr/length(lost)
lost_trt <- lost_trt/length(lost)

## Find common scale for gained/lost
gained_scale <- c(0, max(gained_ctr, gained_trt))
lost_scale <- c(0, max(lost_ctr, lost_trt))
```

Then we can visualize the gained loops in control
and treatment:

```{r, message=FALSE, fig.show="hold", out.width="50%"}
plotMatrix(gained_ctr, zrange=gained_scale)
plotMatrix(gained_trt, zrange=gained_scale)
```

And the lost loops in control and treatment:

```{r, message=FALSE, fig.show="hold", out.width="50%"}
plotMatrix(lost_ctr, zrange=lost_scale)
plotMatrix(lost_trt, zrange=lost_scale)
```

In the [final section of this workshop](vis_with_pg) we will take
these plots and visualize them on a single page with labels.

`mariner` includes the `pileupPixels()` function, which wraps several
of these steps together to simply your workflow. For example,
the code below can be used to accomplish the same task.

```{r, message=FALSE}
## Identify gained/lost loops
res <- rowData(pixels)
gained <- pixels[which(res$padj <= 0.1 & res$log2FoldChange > 0)]
lost <- pixels[which(res$padj <= 0.1 & res$log2FoldChange < 0)]

## Calculate apa matrices
gained_ctr <- pileupPixels(
  x=interactions(gained),
  files=hicFiles[1:6],
  binSize=5e3
)
gained_trt <- pileupPixels(
  x=interactions(gained),
  files=hicFiles[7:12],
  binSize=5e3,
  minPairDist=50e3
)
lost_ctr <- pileupPixels(
  x=interactions(lost),
  files=hicFiles[1:6],
  binSize=5e3
)
lost_trt <- pileupPixels(
  x=interactions(lost),
  files=hicFiles[7:12],
  binSize=5e3
)

## Find common scale for gained/lost
gained_scale <- c(0, max(gained_ctr, gained_trt))
lost_scale <- c(0, max(lost_ctr, lost_trt))
```

Then we can visualize the gained loops in control
and treatment:

```{r, message=FALSE, fig.show="hold", out.width="50%"}
plotMatrix(gained_ctr, zrange=gained_scale)
plotMatrix(gained_trt, zrange=gained_scale)
```

And the lost loops in control and treatment:

```{r, message=FALSE, fig.show="hold", out.width="50%"}
plotMatrix(lost_ctr, zrange=lost_scale)
plotMatrix(lost_trt, zrange=lost_scale)
```

Running each step separately allows you to have the
maximum amount of control over the settings which improves
efficiency and performance. It also gives you more flexibility
to allow custom functions that haven't been explicitly created.

### Aggregate domain analysis of differential loops

In addition to pileups of pixels, we can also perform
pileup analysis for domains. In this analysis, we will
aggregate the whole domain encompassed by the differential
loops so we can visualize how the domains are changing.

We can build a helper function that makes TAD with a 
buffer on either side from the loops.

```{r}
## Write a helper function to convert
## loops into TADs
makeTadsFromLoops <- function(x) {
  d <- InteractionSet::pairdist(x, 'span')/2
  df <- data.frame(
    seqnames1(x),
    start1(x)-d,
    end2(x)+d
  )
  as_ginteractions(cbind(df, df))
}

tads <- makeTadsFromLoops(mergedLoops)
```

Then we can filter these tads for those that are gained
after 72 hours of PMA treatment.

```{r}
## Identify gained TADs
gained <- tads[which(res$padj <= 0.1 & res$log2FoldChange > 0)]
```

The `pullHicMatrices()` function can be used to 
extract count matrices for these tads at a given
resolution (in this case 5 Kb).

```{r}
## Pull TAD regions
mats <- pullHicMatrices(
  x=gained,
  files=hicFiles[7:12],
  binSize=5e3,
  blockSize=100e6
)
```

Because each region is a different size, an
irregular or jagged array object is returned:
  
```{r}
mats
```

The `counts()` function gives access to
these count matrices, which can be realized
with `as.list()` if needed.

```{r}
counts(mats)
```

For an aggregate TAD analysis, we need to
make these irregular arrays the same size.
The `regularize()` function resizes each
matrix to the rows and column dimensions
specified in the `ndim` parameter. The
`scale` parameter will scale each matrix
to its total signal so that each matrix
contributes the same relative signal to
the aggregation. The result is an
`InteractionArray` as we've seen in
previous sections.

```{r, message=FALSE}
## Convert to regular array
reg <- regularize(
  x=mats,
  ndim=c(100, 100),
  nBlocks=10,
  scale=TRUE
)
```

The regularized array can then be
aggregated and visualized like before:

```{r, message=FALSE}
## Aggregate
agg <- aggHicMatrices(reg)

## Visualize
plotMatrix(agg, zrange=c(0, 0.025))
```

We can use the wrapper function `pileupDomains()` to carry
out these steps for us:

```{r, message=FALSE, warning=FALSE, fig.show="hold", out.width="50%"}
## Identify gained TADs
gained <- tads[which(res$padj <= 0.1 & res$log2FoldChange > 0)]

## Calculate aggregate TAD plots
gained_ctr_tads <- pileupDomains(
  x=gained,
  files=hicFiles[1],
  binSize=5e3,
  buffer=0.25
)
gained_trt_tads <- pileupDomains(
  x=gained,
  files=hicFiles[7],
  binSize=5e3,
  buffer=0.25
)

## Find common scale for gained/lost
gq <- quantile(c(gained_ctr_tads, gained_trt_tads), 0.90)
gained_tad_scale <- c(0, gq)

## Visualize
plotMatrix(gained_ctr_tads, zrange=gained_tad_scale)
plotMatrix(gained_trt_tads, zrange=gained_tad_scale)
```

## Visualizing with `plotgardener` <a id="vis_with_pg" />

`plotgardener` is a genomic data visualization package for R.
It uses a coordinate-based plotting system to give users
precise control over the size and placement of visualizations.
In this section we will use `plotgardener` to programmatically
place our APA plots on a page and add text labels.

To visualize multiple plots with `plotgardener` you start by
creating a page:
```{r, echo=FALSE}
knitr::opts_knit$set(global.device = TRUE)
```

```{r, message=FALSE}
library(plotgardener)

## Create page
pageCreate(width=3, height=3)
```

We can pass additional parameters to the `plotMatrix()` function
specifying where we would like these to be placed on the page.

```{r, message=FALSE, results='hold'}
## APA plots
gainedApa <- plotMatrix(
  data=gained_ctr,
  zrange=gained_scale,
  x=0.5,
  y=0.5,
  width=0.95,
  height=0.95,
)
plotMatrix(
  data=gained_trt,
  zrange=gained_scale,
  x=0.5,
  y=1.5,
  width=0.95,
  height=0.95,
)
lostApa <- plotMatrix(
  data=lost_ctr,
  zrange=lost_scale,
  x=1.5,
  y=0.5,
  width=0.95,
  height=0.95,
)
plotMatrix(
  data=lost_trt,
  zrange=lost_scale,
  x=1.5,
  y=1.5,
  width=0.95,
  height=0.95,
)
```

We can use annotation functions to add scale legends. 

```{r message=FALSE, results='hold'}
## APA heatmap legends
annoHeatmapLegend(
  plot=gainedApa,
  orientation='h',
  x=0.5,
  y=2.5,
  width=0.95,
  height=0.1
)
annoHeatmapLegend(
  plot=lostApa,
  orientation='h',
  x=1.5,
  y=2.5,
  width=0.95,
  height=0.1
)
```

and text labels:

```{r, message=FALSE, results='hold'}
## Plot text labels
plotText(
  label="72hr-loops",
  x=0.975,
  y=0.45,
  just=c("center", "bottom")
)
plotText(
  label="0hr-loops",
  x=1.975,
  y=0.45,
  just=c("center", "bottom")
)
plotText(
  label="0hr-hic",
  x=0.45,
  y=0.975,
  rot=90,
  just=c("center", "bottom")
)
plotText(
  label="72hr-hic",
  x=0.45,
  y=1.975,
  rot=90,
  just=c("center", "bottom")
)
```

When the plot is to your liking, you can
remove the guides on the page:

```{r, message=FALSE, results='hold'}
pageGuideHide()
```

`plotgardener` is a very useful tool for making
publication-quality genomic figures. Visit the
[website](https://phanstiellab.github.io/plotgardener/)
to learn more about what you can do with `plotgardener`.

For example, we could plot an example loop beside these
APA plots:

```{r, message=FALSE}
## Create page
pageCreate(width=6, height=3, showGuides=FALSE)

## APA plots
gainedApa <- plotMatrix(
  data=gained_ctr,
  zrange=gained_scale,
  x=0.5,
  y=0.5,
  width=0.95,
  height=0.95,
)
plotMatrix(
  data=gained_trt,
  zrange=gained_scale,
  x=0.5,
  y=1.5,
  width=0.95,
  height=0.95,
)
lostApa <- plotMatrix(
  data=lost_ctr,
  zrange=lost_scale,
  x=1.5,
  y=0.5,
  width=0.95,
  height=0.95,
)
plotMatrix(
  data=lost_trt,
  zrange=lost_scale,
  x=1.5,
  y=1.5,
  width=0.95,
  height=0.95,
)

## APA heatmap legends
annoHeatmapLegend(
  plot=gainedApa,
  orientation='h',
  x=0.5,
  y=2.5,
  width=0.95,
  height=0.1
)
annoHeatmapLegend(
  plot=lostApa,
  orientation='h',
  x=1.5,
  y=2.5,
  width=0.95,
  height=0.1
)

## Plot text labels
plotText(
  label="72hr-loops",
  x=0.975,
  y=0.45,
  just=c("center", "bottom")
)
plotText(
  label="0hr-loops",
  x=1.975,
  y=0.45,
  just=c("center", "bottom")
)
plotText(
  label="0hr-hic",
  x=0.45,
  y=0.975,
  rot=90,
  just=c("center", "bottom")
)
plotText(
  label="72hr-hic",
  x=0.45,
  y=1.975,
  rot=90,
  just=c("center", "bottom")
)

## Plot Hi-C example loop
hicPlot <- plotHicSquare(
  ## Path to .hic file
  data=hicFiles[1],
  
  ## Loci
  chrom="chr1",
  chromstart=213.9e6,
  chromend=215e6,
  
  ## Hi-C params
  resolution=10e3,
  zrange=c(0, 40),
  norm="KR",
  half="top",
  
  ## Placement on page
  x=3,
  y=0.5,
  width=2,
  height=2
)
plotHicSquare(
  ## Path to .hic file
  data=hicFiles[7],
  
  ## Loci
  chrom="chr1",
  chromstart=213.9e6,
  chromend=215e6,
  
  ## Plotting params
  resolution=10e3,
  zrange=c(0, 40),
  norm="KR",
  half="bottom",
  
  ## Placement on page
  x=3,
  y=0.5,
  width=2,
  height=2
)

## Genome coordinate label
annoGenomeLabel(
  plot=hicPlot,
  scale="Mb",
  x=3,
  y=2.5
)

## Hi-C heatmap legend
annoHeatmapLegend(
  plot=hicPlot,
  orientation='v',
  x=5.05,
  y=0.5,
  width=0.1,
  height=1
)

## Plot title & half labels
plotText(
  label="Example Loop; 10Kb bins",
  x=4,
  y=0.45,
  just=c("center", "bottom")
)
plotText(
  label="0hrs",
  x=3.05,
  y=0.55,
  just=c("left", "top")
)
plotText(
  label="72hrs",
  x=4.95,
  y=2.45,
  just=c("right", "bottom")
)
```

## References