<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mariner: Explore the Hi-Cs • Bioc2023mariner</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mariner: Explore the Hi-Cs">
<meta property="og:description" content="Bioc2023mariner">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2023mariner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Bioc2023mariner.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/EricSDavis/Bioc2023mariner" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mariner: Explore the Hi-Cs</h1>
                        <h4 data-toc-skip class="author">Eric S. Davis,
Sarah M. Parker, Manjari Kiran, Nicole Kramer, Douglas H. Phanstiel</h4>
            
            <h4 data-toc-skip class="date">2023-07-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ericsdavis/Bioc2023mariner/blob/HEAD/vignettes/Bioc2023mariner.Rmd" class="external-link"><code>vignettes/Bioc2023mariner.Rmd</code></a></small>
      <div class="hidden name"><code>Bioc2023mariner.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>3D chromatin structure plays an integral, yet incompletely understood
role in the long-distance regulation of genes by enhancers and
repressors. Disruption or aberrant formation of these long-range
interactions can result in developmental abnormalities and diseases,
such as cancer. Therefore, deriving biological insights from 3D
chromatin structure experiments, such as Hi-C or Micro-C, is essential
for understanding and correcting human disease.</p>
<p>In this workshop, we will use a recently published dataset <span class="citation">(<a href="#ref-Bond2022.10.31.514600" role="doc-biblioref">Bond et al. 2022</a>)</span> exploring chromatin
loop dynamics during a megakaryocyte differentiation event to
demonstrate a differential looping analysis with <code>mariner</code>,
<code>DESeq2</code>, and <code>plotgardener</code>. Specifically, we
will show how <code>mariner</code> can be used to merge redundant loop
calls and extract Hi-C contact frequency between putative loop anchors
to use with <code>DESeq2</code> to identify differential loops. We will
then conduct pileup analyses of Hi-C pixels, domains, and boundaries
with <code>mariner</code>. Finally, we will demonstrate how these
analyses can be visualized together in a publication-ready multi-panel
figure with <code>plotgardener</code>.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the <code>GenomicRanges</code> and
<code>InteractionSet</code> classes</li>
</ul>
</div>
<div class="section level3">
<h3 id="optional-background-reading">Optional background reading<a class="anchor" aria-label="anchor" href="#optional-background-reading"></a>
</h3>
<ul>
<li>Biological background
<ul>
<li>Hi-C background <span class="citation">(<a href="#ref-rao_3d_2014" role="doc-biblioref">Rao et al. 2014</a>)</span>
</li>
<li>Dataset used in workshop <span class="citation">(<a href="#ref-Bond2022.10.31.514600" role="doc-biblioref">Bond et al.
2022</a>)</span>
</li>
</ul>
</li>
<li>Other <em>Bioconductor</em> packages
<ul>
<li>
<code>InteractionSet</code> <span class="citation">(<a href="#ref-InteractionSet" role="doc-biblioref">Lun, Perry, and
Ing-Simmons 2016</a>)</span>
</li>
<li>
<code>DelayedArray</code> <span class="citation">(<a href="#ref-DelayedArray" role="doc-biblioref">Pagès
2023</a>)</span>
</li>
<li>
<code>DESeq2</code> <span class="citation">(<a href="#ref-DESeq2" role="doc-biblioref">Love, Huber, and Anders 2014</a>; <a href="#ref-apeglm" role="doc-biblioref">Zhu, Ibrahim, and Love
2018</a>)</span>
</li>
<li>
<code>plotgardener</code> <span class="citation">(<a href="#ref-plotgardener" role="doc-biblioref">Kramer et al.
2022</a>)</span>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>The format is lecture + lab, with time for data exploration.</p>
</div>
<div class="section level3">
<h3 id="major-r-bioconductor-packages-used">Major <em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#major-r-bioconductor-packages-used"></a>
</h3>
<ul>
<li><code>mariner</code></li>
<li><code>DESeq2</code></li>
<li><code>plotgardener</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>90-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview of Hi-C data</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Workshop datasets</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>Differential analysis with <code>DESeq2</code>
</td>
<td>25m</td>
</tr>
<tr class="even">
<td>Aggregating interactions</td>
<td>25m</td>
</tr>
<tr class="odd">
<td>Visualization with <code>plotgardener</code>
</td>
<td>25m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
</div>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Understand how to work with interaction data in R.</li>
<li>Learn which <code>mariner</code> functions to use for extracting
Hi-C data.</li>
<li>How to use accessors to get information from <code>mariner</code>
objects.</li>
<li>Integrating <code>mariner</code> with <code>plotgardener</code> for
programmatically visualizing genomic data.</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Conduct differential interaction analysis with <code>mariner</code>
and <code>DESeq2</code>.</li>
<li>Make aggregate plots with <code>mariner</code>.</li>
<li>Visualize aggregate plots and genomic regions with
<code>plotgardener</code>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="overview-of-hi-c-data">Overview of Hi-C data<a class="anchor" aria-label="anchor" href="#overview-of-hi-c-data"></a>
</h2>
<iframe src="https://ericscottdavis.com/HowToHic" style="border: none; width: 100%; height: 500px;">
</iframe>
<div class="section level3">
<h3 id="terminology">Terminology<a class="anchor" aria-label="anchor" href="#terminology"></a>
</h3>
<ul>
<li><p><em>anchors</em>: the genomic loci (genomic ranges) that make up
an <em>interaction</em>.</p></li>
<li><p><em>interactions</em>: a pair of genomic loci (genomic
ranges).</p></li>
<li><p><em>pixel</em>: a two-dimensional bin that represents the
<em>interaction</em> between two fixed-width genomic loci.</p></li>
<li><p><em>contact frequency</em>: the number of Hi-C reads assigned to
a pixel.</p></li>
<li><p><em>loop</em>: a <em>pixel</em> that is enriched for <em>contact
frequency</em> compared to its local background. Biologically, a
<em>loop</em> represents a point of frequent or sustained contact
between genomic loci.</p></li>
<li><p><em>the diagonal</em>: points of self-interaction along the Hi-C
matrix.</p></li>
<li><p><em>upper triangular</em>: the portion of the Hi-C matrix above
<em>the diagonal</em>. <em>Interactions</em> where the first
<em>anchor</em> is less than the second <em>anchor</em> appear
here.</p></li>
<li><p><em>lower triangular</em>: the portion of the Hi-C matrix below
<em>the diagonal</em>. <em>Interactions</em> where the first
<em>anchor</em> is greater than the second <em>anchor</em> appear
here.</p></li>
<li><p><em>resolution</em>: the ability to distinguish between two
neighboring points on a Hi-C matrix. The higher the <em>resolution</em>
the more precise the contact between pairs of loci.</p></li>
<li><p><em>binSize</em>: Hi-C reads are quantified along the genome in
regular bins. The size of these bins determines the <em>resolution</em>
of the Hi-C matrix. Larger <em>bins</em> lead to lower
<em>resolution</em> while smaller <em>bins</em> mean higher
<em>resolution</em>. Since the quantification of reads into bins is time
consuming, <code>.hic</code> files are processed at several
pre-determined <code>binSizes</code> (also called
<em>resolutions</em>).</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workshop-datasets">Workshop datasets<a class="anchor" aria-label="anchor" href="#workshop-datasets"></a>
</h2>
<p>The dataset used in this workshop comes from one of our recently
accepted papers, led by the talented Marielle Bond, where we explore
chromatin loop dynamics during megakaryocyte differentiation <span class="citation">(<a href="#ref-Bond2022.10.31.514600" role="doc-biblioref">Bond et al. 2022</a>)</span>. In short, deeply
sequenced Hi-C was conducted in K562 cells before and after treatment
with PMA for 0, 6, or 72 hours. RNA-seq, ATAC-seq, and CUT&amp;RUN for
H3K27ac was also performed at each timepoint, but will not be used in
this workshop.</p>
<p><img src="img/mega_slide.png" width="75%" style="display: block; margin: auto;"></p>
<p>The Hi-C data in this study was processed using a modified version of
the <code>Juicer</code> pipeline <span class="citation">(<a href="#ref-durand_juicer_2016" role="doc-biblioref">Durand et al.
2016</a>; <a href="#ref-eric_davis_2022_7514346" role="doc-biblioref">Davis and Reed 2022</a>)</span> resulting in
<code>.hic</code> files for each biological and technical replicate.
Loops were identified using <a href="https://github.com/PouletAxel/SIP" class="external-link"><code>SIP</code></a> <span class="citation">(<a href="#ref-Rowley01032020" role="doc-biblioref">Rowley et al. 2020</a>)</span> for each
timepoint.</p>
<p>Since the full dataset is very large, we will be working with a
subset of these data in the workshop. We will use 6 Hi-C replicates
(ignoring biological/technical designations) for the control (0 hours of
PMA) and 6 replicates for the treatment (72 hours of PMA). The
<code>.hic</code> files and loop files have been filtered to only
contain reads from the first chromosome of the human hg38 genome
build.</p>
<p>The following shows the directory structure of the files used in this
workshop:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vignettes/data/</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> hic</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_1_1_chr1.hic"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_1_2_chr1.hic"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_2_1_chr1.hic"</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_2_2_chr1.hic"</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_3_1_chr1.hic"</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_3_2_chr1.hic"</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"72hr_1_1_chr1.hic"</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"72hr_1_2_chr1.hic"</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"72hr_2_1_chr1.hic"</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"72hr_2_2_chr1.hic"</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"72hr_3_1_chr1.hic"</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   └── <span class="st">"72hr_3_2_chr1.hic"</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> loops</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"0hr_5kbLoops_chr1.txt"</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   ├── <span class="st">"6hr_5kbLoops_chr1.txt"</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>   └── <span class="st">"72hr_5kbLoops_chr1.txt"</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> samples</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> <span class="st">"samplesheet.txt"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-analysis-with-deseq2">Differential analysis with <code>DESeq2</code><a class="anchor" aria-label="anchor" href="#differential-analysis-with-deseq2"></a>
</h2>
<p>In this section we will use <code>DESeq2</code> and
<code>mariner</code> to identify differential loops between K562 cells
treated with PMA for 0 or 72 hours.</p>
<div class="section level3">
<h3 id="merge-redundant-loops">Merge redundant loops<a class="anchor" aria-label="anchor" href="#merge-redundant-loops"></a>
</h3>
<p>The first step in this analysis is to read in and combine loop calls
for K562 cells treated for 0, 6, or 72 hours. Due to technical and
biological variability, loop callers can identify slightly different
pixels as loops in different datasets. Sometimes this represents a true
biological shift in the position of the loop anchors, but more often is
due to noise and technical variability. Furthermore, testing these
redundant interactions reduces our power to identify true differences in
looping.</p>
<p><img src="img/cluster_merge.png" width="75%" style="display: block; margin: auto;"></p>
<p>To combat this effect, we can use <code>mariner</code> functions to
cluster then combine loops between conditions. First, we start by
reading in BEDPE-formatted loops, and converting them to
<code>GInteractions</code> objects.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ericscottdavis.com/mariner/" class="external-link">mariner</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Read in loops</span></span>
<span><span class="va">loops00hr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/loops/0hr_5kbLoops_chr1.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">loops06hr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/loops/6hr_5kbLoops_chr1.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">loops72hr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/loops/72hr_5kbLoops_chr1.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Put into a list</span></span>
<span><span class="va">loopList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"0hr"</span><span class="op">=</span><span class="va">loops00hr</span>, <span class="st">"6hr"</span><span class="op">=</span><span class="va">loops06hr</span>, <span class="st">"72hr"</span><span class="op">=</span><span class="va">loops72hr</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Convert to GInteractions objects</span></span>
<span><span class="va">loopList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">loopList</span>, <span class="va">as_ginteractions</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Show list of GInteractions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">loopList</span>, <span class="va">summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`0hr`</span></span>
<span><span class="co">#&gt; [1] "GInteractions object of length 1157 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`6hr`</span></span>
<span><span class="co">#&gt; [1] "GInteractions object of length 1083 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`72hr`</span></span>
<span><span class="co">#&gt; [1] "GInteractions object of length 1468 with 9 metadata columns"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loopList</span><span class="op">[[</span><span class="st">"0hr"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; GInteractions object with 1157 interactions and 9 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;      [1]      chr1 178540000-178545000 ---      chr1 179000000-179005000 |</span></span>
<span><span class="co">#&gt;      [2]      chr1   47175000-47180000 ---      chr1   47705000-47710000 |</span></span>
<span><span class="co">#&gt;      [3]      chr1 100275000-100280000 ---      chr1 100345000-100350000 |</span></span>
<span><span class="co">#&gt;      [4]      chr1 111030000-111035000 ---      chr1 111390000-111395000 |</span></span>
<span><span class="co">#&gt;      [5]      chr1 172740000-172745000 ---      chr1 173415000-173420000 |</span></span>
<span><span class="co">#&gt;      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [1153]      chr1 152050000-152055000 ---      chr1 152185000-152190000 |</span></span>
<span><span class="co">#&gt;   [1154]      chr1 198095000-198100000 ---      chr1 198685000-198690000 |</span></span>
<span><span class="co">#&gt;   [1155]      chr1 174415000-174420000 ---      chr1 174720000-174725000 |</span></span>
<span><span class="co">#&gt;   [1156]      chr1 176560000-176565000 ---      chr1 176880000-176885000 |</span></span>
<span><span class="co">#&gt;   [1157]      chr1   64680000-64685000 ---      chr1   65035000-65040000 |</span></span>
<span><span class="co">#&gt;                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt;          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]       0,0,0    3.73373                0.979253       2.46663</span></span>
<span><span class="co">#&gt;      [2]       0,0,0    3.17268                1.000000       1.88925</span></span>
<span><span class="co">#&gt;      [3]       0,0,0    1.92717                0.910580       1.56251</span></span>
<span><span class="co">#&gt;      [4]       0,0,0    2.62830                0.967662       1.87664</span></span>
<span><span class="co">#&gt;      [5]       0,0,0    3.26733                0.988218       2.07950</span></span>
<span><span class="co">#&gt;      ...         ...        ...                     ...           ...</span></span>
<span><span class="co">#&gt;   [1153]       0,0,0    5.32936                1.000000       3.19002</span></span>
<span><span class="co">#&gt;   [1154]       0,0,0    2.34929                0.956688       1.45957</span></span>
<span><span class="co">#&gt;   [1155]       0,0,0    2.44098                0.952570       1.75091</span></span>
<span><span class="co">#&gt;   [1156]       0,0,0    4.00059                0.993421       2.89171</span></span>
<span><span class="co">#&gt;   [1157]       0,0,0    2.44518                0.914732       1.70571</span></span>
<span><span class="co">#&gt;          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std</span></span>
<span><span class="co">#&gt;                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]               1.109284               1.500514   2.88972  0.526448</span></span>
<span><span class="co">#&gt;      [2]               1.048126               1.412119   2.13157  0.498198</span></span>
<span><span class="co">#&gt;      [3]               0.372410               0.527703   2.08363  0.286243</span></span>
<span><span class="co">#&gt;      [4]               0.924161               1.071454   2.61039  0.453281</span></span>
<span><span class="co">#&gt;      [5]               1.119304               1.963420   3.44674  0.564086</span></span>
<span><span class="co">#&gt;      ...                    ...                    ...       ...       ...</span></span>
<span><span class="co">#&gt;   [1153]               1.571963               2.439878   3.74575  0.865201</span></span>
<span><span class="co">#&gt;   [1154]               1.003424               1.356165   2.24773  0.492896</span></span>
<span><span class="co">#&gt;   [1155]               0.604154               1.044648   2.51178  0.312849</span></span>
<span><span class="co">#&gt;   [1156]               0.713450               1.819302   4.39018  0.372197</span></span>
<span><span class="co">#&gt;   [1157]               0.726476               0.800235   1.81645  0.427180</span></span>
<span><span class="co">#&gt;              value</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]   3.87575</span></span>
<span><span class="co">#&gt;      [2]   3.06324</span></span>
<span><span class="co">#&gt;      [3]   2.41466</span></span>
<span><span class="co">#&gt;      [4]   3.43187</span></span>
<span><span class="co">#&gt;      [5]   4.44168</span></span>
<span><span class="co">#&gt;      ...       ...</span></span>
<span><span class="co">#&gt;   [1153]   5.14305</span></span>
<span><span class="co">#&gt;   [1154]   3.13966</span></span>
<span><span class="co">#&gt;   [1155]   3.04881</span></span>
<span><span class="co">#&gt;   [1156]   5.02436</span></span>
<span><span class="co">#&gt;   [1157]   2.46221</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 1988 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Then we can use the <code><a href="http://ericscottdavis.com/mariner/reference/mergePairs.html" class="external-link">mergePairs()</a></code> function to cluster
then merge the list of <code>GInteraction</code> objects.</p>
<p>This function uses <code>dbscan</code> to cluster loops by a given
genomic <code>radius</code> in base-pairs. It selects a representative
interaction for each cluster using either the metadata
<code>column</code> provided, or if <code>column=NULL</code> it
calculates new ranges for each pair by calculating the median of
modes.</p>
<p><img src="img/cluster_merge2.png" width="50%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mergedLoops</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/mergePairs.html" class="external-link">mergePairs</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">loopList</span>, radius<span class="op">=</span><span class="fl">10e3</span>, column<span class="op">=</span><span class="st">"APScoreAvg"</span><span class="op">)</span></span>
<span><span class="va">mergedLoops</span></span>
<span><span class="co">#&gt; MergedGInteractions object with 2339 interactions and 9 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;      [1]      chr1 100275000-100280000 ---      chr1 100345000-100350000 |</span></span>
<span><span class="co">#&gt;      [2]      chr1   92855000-92860000 ---      chr1   93060000-93065000 |</span></span>
<span><span class="co">#&gt;      [3]      chr1   77975000-77980000 ---      chr1   78110000-78115000 |</span></span>
<span><span class="co">#&gt;      [4]      chr1   89500000-89505000 ---      chr1   89600000-89605000 |</span></span>
<span><span class="co">#&gt;      [5]      chr1 213005000-213010000 ---      chr1 213070000-213075000 |</span></span>
<span><span class="co">#&gt;      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [2335]      chr1 148135000-148140000 ---      chr1 148250000-148255000 |</span></span>
<span><span class="co">#&gt;   [2336]      chr1 197520000-197525000 ---      chr1 198685000-198690000 |</span></span>
<span><span class="co">#&gt;   [2337]      chr1 235515000-235520000 ---      chr1 235595000-235600000 |</span></span>
<span><span class="co">#&gt;   [2338]      chr1 203040000-203045000 ---      chr1 203335000-203340000 |</span></span>
<span><span class="co">#&gt;   [2339]      chr1 172450000-172455000 ---      chr1 172720000-172725000 |</span></span>
<span><span class="co">#&gt;                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt;          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]       0,0,0    1.92717                0.910580       1.56251</span></span>
<span><span class="co">#&gt;      [2]       0,0,0    3.34108                0.965907       2.43196</span></span>
<span><span class="co">#&gt;      [3]       0,0,0    3.20562                1.000000       2.14776</span></span>
<span><span class="co">#&gt;      [4]       0,0,0    2.80295                1.000000       2.37292</span></span>
<span><span class="co">#&gt;      [5]       0,0,0    2.25636                1.000000       1.65880</span></span>
<span><span class="co">#&gt;      ...         ...        ...                     ...           ...</span></span>
<span><span class="co">#&gt;   [2335]       0,0,0    3.64734                0.985123       2.10353</span></span>
<span><span class="co">#&gt;   [2336]       0,0,0    3.68298                0.995663       1.72712</span></span>
<span><span class="co">#&gt;   [2337]       0,0,0    2.18036                1.000000       1.83207</span></span>
<span><span class="co">#&gt;   [2338]       0,0,0    3.97304                0.983620       2.38127</span></span>
<span><span class="co">#&gt;   [2339]       0,0,0    5.15098                1.000000       3.78385</span></span>
<span><span class="co">#&gt;          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std</span></span>
<span><span class="co">#&gt;                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]               0.372410               0.527703   2.08363  0.286243</span></span>
<span><span class="co">#&gt;      [2]               0.779426               1.046919   2.68620  0.308547</span></span>
<span><span class="co">#&gt;      [3]               0.760126               1.022160   2.04379  0.340562</span></span>
<span><span class="co">#&gt;      [4]               0.287866               0.472579   2.31595  0.153064</span></span>
<span><span class="co">#&gt;      [5]               0.606417               0.617856   1.68683  0.272008</span></span>
<span><span class="co">#&gt;      ...                    ...                    ...       ...       ...</span></span>
<span><span class="co">#&gt;   [2335]               1.700268               1.933103   2.69700  0.779928</span></span>
<span><span class="co">#&gt;   [2336]               2.983473               4.808131   4.94031  1.566966</span></span>
<span><span class="co">#&gt;   [2337]               0.209327               0.392278   1.80495  0.101174</span></span>
<span><span class="co">#&gt;   [2338]               1.612138               1.768132   2.67911  0.721396</span></span>
<span><span class="co">#&gt;   [2339]               0.871347               1.395824   3.64185  0.561411</span></span>
<span><span class="co">#&gt;              value</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]   2.41466</span></span>
<span><span class="co">#&gt;      [2]   3.37902</span></span>
<span><span class="co">#&gt;      [3]   2.71946</span></span>
<span><span class="co">#&gt;      [4]   2.57183</span></span>
<span><span class="co">#&gt;      [5]   2.22587</span></span>
<span><span class="co">#&gt;      ...       ...</span></span>
<span><span class="co">#&gt;   [2335]   4.20835</span></span>
<span><span class="co">#&gt;   [2336]   7.59229</span></span>
<span><span class="co">#&gt;   [2337]   1.99102</span></span>
<span><span class="co">#&gt;   [2338]   4.11212</span></span>
<span><span class="co">#&gt;   [2339]   4.41638</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3596 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="exercise---finding-de-novo-loops">Exercise - Finding de novo loops<a class="anchor" aria-label="anchor" href="#exercise---finding-de-novo-loops"></a>
</h3>
<p>Let’s take a closer look at the <code>MergedGInteractions</code>
object. This object looks and behaves as a normal
<code>GInteractions</code> object, but also contains slots storing the
clustered interactions and associated metadata. There are a few key
accessor functions that allow you to extract, aggregate, and subset the
interactions according to how they have clustered.</p>
<p>The input to <code><a href="http://ericscottdavis.com/mariner/reference/mergePairs.html" class="external-link">mergePairs()</a></code> was a list of
<code>GInteractions</code> objects. The <code><a href="http://ericscottdavis.com/mariner/reference/sources.html" class="external-link">sources()</a></code> accessor
returns the names of each input source file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/sources.html" class="external-link">sources</a></span><span class="op">(</span><span class="va">mergedLoops</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "0hr"  "6hr"  "72hr"</span></span></code></pre></div>
<p>Each row of the <code>MergedGInteractions</code> object represents
the selected interaction from each cluster of one or more interactions.
To view all interactions in a cluster, you can use the
<code><a href="http://ericscottdavis.com/mariner/reference/clusters.html" class="external-link">clusters()</a></code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/clusters.html" class="external-link">clusters</a></span><span class="op">(</span><span class="va">mergedLoops</span><span class="op">[</span><span class="fl">1500</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;    seqnames1   start1     end1 width1 strand1 seqnames2   start2     end2</span></span>
<span><span class="co">#&gt; 1:      chr1 94515000 94520000   5001       *      chr1 94795000 94800000</span></span>
<span><span class="co">#&gt; 2:      chr1 94510000 94515000   5001       *      chr1 94805000 94810000</span></span>
<span><span class="co">#&gt; 3:      chr1 94515000 94520000   5001       *      chr1 94800000 94805000</span></span>
<span><span class="co">#&gt;    width2 strand2 color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt; 1:   5001       * 0,0,0   4.460383               0.9974052      2.909975</span></span>
<span><span class="co">#&gt; 2:   5001       * 0,0,0   4.523593               0.9974524      2.931985</span></span>
<span><span class="co">#&gt; 3:   5001       * 0,0,0   5.669824               0.9997473      3.897449</span></span>
<span><span class="co">#&gt;    Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2      avg       std   value</span></span>
<span><span class="co">#&gt; 1:               2.027463               2.220452 4.152661 0.9445974 5.95485</span></span>
<span><span class="co">#&gt; 2:               1.864137               2.351764 4.316209 1.2186011 5.97322</span></span>
<span><span class="co">#&gt; 3:               1.643575               3.224509 6.823265 1.2357823 8.28422</span></span>
<span><span class="co">#&gt;     src</span></span>
<span><span class="co">#&gt; 1:  0hr</span></span>
<span><span class="co">#&gt; 2:  6hr</span></span>
<span><span class="co">#&gt; 3: 72hr</span></span></code></pre></div>
<p>Notice that it returns a <code>data.table</code> with a column
appended to the end, <code>src</code>, indicating the source file of
each interaction. All of the metadata for each specific interaction is
also included. If you want to summarize one or more of these values for
each cluster in the object, the <code><a href="http://ericscottdavis.com/mariner/reference/aggMetadata.html" class="external-link">aggMetadata()</a></code> function is
handy.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/aggMetadata.html" class="external-link">aggMetadata</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mergedLoops</span>, columns<span class="op">=</span><span class="st">"APScoreAvg"</span>, funs<span class="op">=</span><span class="st">"mean"</span><span class="op">)</span></span>
<span><span class="co">#&gt; MergedGInteractions object with 2339 interactions and 10 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;      [1]      chr1 100275000-100280000 ---      chr1 100345000-100350000 |</span></span>
<span><span class="co">#&gt;      [2]      chr1   92855000-92860000 ---      chr1   93060000-93065000 |</span></span>
<span><span class="co">#&gt;      [3]      chr1   77975000-77980000 ---      chr1   78110000-78115000 |</span></span>
<span><span class="co">#&gt;      [4]      chr1   89500000-89505000 ---      chr1   89600000-89605000 |</span></span>
<span><span class="co">#&gt;      [5]      chr1 213005000-213010000 ---      chr1 213070000-213075000 |</span></span>
<span><span class="co">#&gt;      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [2335]      chr1 148135000-148140000 ---      chr1 148250000-148255000 |</span></span>
<span><span class="co">#&gt;   [2336]      chr1 197520000-197525000 ---      chr1 198685000-198690000 |</span></span>
<span><span class="co">#&gt;   [2337]      chr1 235515000-235520000 ---      chr1 235595000-235600000 |</span></span>
<span><span class="co">#&gt;   [2338]      chr1 203040000-203045000 ---      chr1 203335000-203340000 |</span></span>
<span><span class="co">#&gt;   [2339]      chr1 172450000-172455000 ---      chr1 172720000-172725000 |</span></span>
<span><span class="co">#&gt;                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt;          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]       0,0,0    1.92717                0.910580       1.56251</span></span>
<span><span class="co">#&gt;      [2]       0,0,0    3.34108                0.965907       2.43196</span></span>
<span><span class="co">#&gt;      [3]       0,0,0    3.20562                1.000000       2.14776</span></span>
<span><span class="co">#&gt;      [4]       0,0,0    2.80295                1.000000       2.37292</span></span>
<span><span class="co">#&gt;      [5]       0,0,0    2.25636                1.000000       1.65880</span></span>
<span><span class="co">#&gt;      ...         ...        ...                     ...           ...</span></span>
<span><span class="co">#&gt;   [2335]       0,0,0    3.64734                0.985123       2.10353</span></span>
<span><span class="co">#&gt;   [2336]       0,0,0    3.68298                0.995663       1.72712</span></span>
<span><span class="co">#&gt;   [2337]       0,0,0    2.18036                1.000000       1.83207</span></span>
<span><span class="co">#&gt;   [2338]       0,0,0    3.97304                0.983620       2.38127</span></span>
<span><span class="co">#&gt;   [2339]       0,0,0    5.15098                1.000000       3.78385</span></span>
<span><span class="co">#&gt;          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std</span></span>
<span><span class="co">#&gt;                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]               0.372410               0.527703   2.08363  0.286243</span></span>
<span><span class="co">#&gt;      [2]               0.779426               1.046919   2.68620  0.308547</span></span>
<span><span class="co">#&gt;      [3]               0.760126               1.022160   2.04379  0.340562</span></span>
<span><span class="co">#&gt;      [4]               0.287866               0.472579   2.31595  0.153064</span></span>
<span><span class="co">#&gt;      [5]               0.606417               0.617856   1.68683  0.272008</span></span>
<span><span class="co">#&gt;      ...                    ...                    ...       ...       ...</span></span>
<span><span class="co">#&gt;   [2335]               1.700268               1.933103   2.69700  0.779928</span></span>
<span><span class="co">#&gt;   [2336]               2.983473               4.808131   4.94031  1.566966</span></span>
<span><span class="co">#&gt;   [2337]               0.209327               0.392278   1.80495  0.101174</span></span>
<span><span class="co">#&gt;   [2338]               1.612138               1.768132   2.67911  0.721396</span></span>
<span><span class="co">#&gt;   [2339]               0.871347               1.395824   3.64185  0.561411</span></span>
<span><span class="co">#&gt;              value mean.APScoreAvg</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]   2.41466         1.92717</span></span>
<span><span class="co">#&gt;      [2]   3.37902         3.34108</span></span>
<span><span class="co">#&gt;      [3]   2.71946         3.20562</span></span>
<span><span class="co">#&gt;      [4]   2.57183         2.80295</span></span>
<span><span class="co">#&gt;      [5]   2.22587         2.25636</span></span>
<span><span class="co">#&gt;      ...       ...             ...</span></span>
<span><span class="co">#&gt;   [2335]   4.20835         3.55909</span></span>
<span><span class="co">#&gt;   [2336]   7.59229         3.08576</span></span>
<span><span class="co">#&gt;   [2337]   1.99102         2.10958</span></span>
<span><span class="co">#&gt;   [2338]   4.11212         3.18861</span></span>
<span><span class="co">#&gt;   [2339]   4.41638         4.64555</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3596 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Finally, we can subset these interactions by the source file using
the <code><a href="http://ericscottdavis.com/mariner/reference/sets.html" class="external-link">sets()</a></code> function. This function includes optional
parameters, <code>include</code> and <code>exclude</code>, which control
which sources should be grouped in the output. By default a list of
<code>MergedGInteractions</code> objects is returned that contains the
set of interactions belonging to each combination of sources.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/sets.html" class="external-link">sets</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mergedLoops</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`0hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 400 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`6hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 357 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`72hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 692 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`0hr_6hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 114 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`0hr_72hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 164 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`6hr_72hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 133 with 9 metadata columns"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`0hr_6hr_72hr`</span></span>
<span><span class="co">#&gt; [1] "MergedGInteractions object of length 479 with 9 metadata columns"</span></span></code></pre></div>
<p>We can visualize these sets with an <code>UpSet</code> plot:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR" class="external-link">UpSetR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Count loops in each set</span></span>
<span><span class="va">loopSets</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/sets.html" class="external-link">sets</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mergedLoops</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">length</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Convert to an expression set for UpSetR</span></span>
<span><span class="va">loopSets</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">loopSets</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">"&amp;"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">loopSets</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromExpression.html" class="external-link">fromExpression</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html" class="external-link">upset</a></span><span class="op">(</span><span class="va">loopSets</span>, order.by<span class="op">=</span><span class="st">"freq"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>If we want to find loops that form <em>de</em> <em>novo</em> after 72
hours explicitly excluding the 0 and 6 hour time points, we can use the
following code:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/sets.html" class="external-link">sets</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mergedLoops</span>, include<span class="op">=</span><span class="st">"72hr"</span>, exclude<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0hr"</span>, <span class="st">"6hr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; MergedGInteractions object with 692 interactions and 9 metadata columns:</span></span>
<span><span class="co">#&gt;         seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;     [1]      chr1   63910000-63915000 ---      chr1   64135000-64140000 |</span></span>
<span><span class="co">#&gt;     [2]      chr1 145405000-145410000 ---      chr1 145720000-145725000 |</span></span>
<span><span class="co">#&gt;     [3]      chr1   25890000-25895000 ---      chr1   26160000-26165000 |</span></span>
<span><span class="co">#&gt;     [4]      chr1   21390000-21395000 ---      chr1   21445000-21450000 |</span></span>
<span><span class="co">#&gt;     [5]      chr1   86420000-86425000 ---      chr1   87080000-87085000 |</span></span>
<span><span class="co">#&gt;     ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [688]      chr1   84505000-84510000 ---      chr1   84595000-84600000 |</span></span>
<span><span class="co">#&gt;   [689]      chr1   52125000-52130000 ---      chr1   52295000-52300000 |</span></span>
<span><span class="co">#&gt;   [690]      chr1 185230000-185235000 ---      chr1 185680000-185685000 |</span></span>
<span><span class="co">#&gt;   [691]      chr1   75260000-75265000 ---      chr1   75580000-75585000 |</span></span>
<span><span class="co">#&gt;   [692]      chr1 168440000-168445000 ---      chr1 169115000-169120000 |</span></span>
<span><span class="co">#&gt;               color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt;         &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]       0,0,0    2.61091                0.982305       1.76666</span></span>
<span><span class="co">#&gt;     [2]       0,0,0    6.50401                1.000000       3.21824</span></span>
<span><span class="co">#&gt;     [3]       0,0,0    2.66133                0.932926       1.66434</span></span>
<span><span class="co">#&gt;     [4]       0,0,0    1.94896                0.905653       1.61320</span></span>
<span><span class="co">#&gt;     [5]       0,0,0    2.52387                0.991184       1.71699</span></span>
<span><span class="co">#&gt;     ...         ...        ...                     ...           ...</span></span>
<span><span class="co">#&gt;   [688]       0,0,0    2.11616                0.915387       1.50892</span></span>
<span><span class="co">#&gt;   [689]       0,0,0    2.97274                0.990751       2.10584</span></span>
<span><span class="co">#&gt;   [690]       0,0,0    2.26386                0.961318       1.68927</span></span>
<span><span class="co">#&gt;   [691]       0,0,0    2.22447                0.977893       1.81455</span></span>
<span><span class="co">#&gt;   [692]       0,0,0    2.45284                0.977047       1.80242</span></span>
<span><span class="co">#&gt;         Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]               1.329581               1.373801   2.85305  0.526063</span></span>
<span><span class="co">#&gt;     [2]               1.847853               1.945681   1.99281  0.857298</span></span>
<span><span class="co">#&gt;     [3]               0.930764               1.116354   1.87489  0.452668</span></span>
<span><span class="co">#&gt;     [4]               0.402415               0.434336   2.00332  0.208011</span></span>
<span><span class="co">#&gt;     [5]               1.211365               1.757942   3.65491  0.486245</span></span>
<span><span class="co">#&gt;     ...                    ...                    ...       ...       ...</span></span>
<span><span class="co">#&gt;   [688]               0.699734               0.757569   1.84793  0.349716</span></span>
<span><span class="co">#&gt;   [689]               1.279305               1.494485   3.54655  0.494399</span></span>
<span><span class="co">#&gt;   [690]               0.845134               0.867392   2.50149  0.534297</span></span>
<span><span class="co">#&gt;   [691]               0.665425               0.764979   3.22076  0.466287</span></span>
<span><span class="co">#&gt;   [692]               0.980279               1.073817   2.90334  0.565814</span></span>
<span><span class="co">#&gt;             value</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]   4.03490</span></span>
<span><span class="co">#&gt;     [2]   3.63535</span></span>
<span><span class="co">#&gt;     [3]   2.70224</span></span>
<span><span class="co">#&gt;     [4]   2.36102</span></span>
<span><span class="co">#&gt;     [5]   4.73168</span></span>
<span><span class="co">#&gt;     ...       ...</span></span>
<span><span class="co">#&gt;   [688]   2.46992</span></span>
<span><span class="co">#&gt;   [689]   4.68371</span></span>
<span><span class="co">#&gt;   [690]   3.25272</span></span>
<span><span class="co">#&gt;   [691]   3.81225</span></span>
<span><span class="co">#&gt;   [692]   3.77470</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3596 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>To find loops that were present at either 6 or 72 hrs, but not
present at 0 hrs, we can exclude the 0hr time point:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/sets.html" class="external-link">sets</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">mergedLoops</span>, exclude<span class="op">=</span><span class="st">"0hr"</span><span class="op">)</span></span>
<span><span class="co">#&gt; MergedGInteractions object with 1182 interactions and 9 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;      [1]      chr1   87325000-87330000 ---      chr1   87435000-87440000 |</span></span>
<span><span class="co">#&gt;      [2]      chr1   21505000-21510000 ---      chr1   21590000-21595000 |</span></span>
<span><span class="co">#&gt;      [3]      chr1   66995000-67000000 ---      chr1   67195000-67200000 |</span></span>
<span><span class="co">#&gt;      [4]      chr1   90720000-90725000 ---      chr1   91160000-91165000 |</span></span>
<span><span class="co">#&gt;      [5]      chr1 205665000-205670000 ---      chr1 205965000-205970000 |</span></span>
<span><span class="co">#&gt;      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [1178]      chr1 148135000-148140000 ---      chr1 148250000-148255000 |</span></span>
<span><span class="co">#&gt;   [1179]      chr1 197520000-197525000 ---      chr1 198685000-198690000 |</span></span>
<span><span class="co">#&gt;   [1180]      chr1 235515000-235520000 ---      chr1 235595000-235600000 |</span></span>
<span><span class="co">#&gt;   [1181]      chr1 203040000-203045000 ---      chr1 203335000-203340000 |</span></span>
<span><span class="co">#&gt;   [1182]      chr1 172450000-172455000 ---      chr1 172720000-172725000 |</span></span>
<span><span class="co">#&gt;                color APScoreAvg ProbabilityofEnrichment RegAPScoreAvg</span></span>
<span><span class="co">#&gt;          &lt;character&gt;  &lt;numeric&gt;               &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]       0,0,0    2.24145                0.922331       1.47099</span></span>
<span><span class="co">#&gt;      [2]       0,0,0    2.00441                0.922253       1.58588</span></span>
<span><span class="co">#&gt;      [3]       0,0,0    2.16722                0.931100       1.39858</span></span>
<span><span class="co">#&gt;      [4]       0,0,0    2.54873                0.988199       1.68114</span></span>
<span><span class="co">#&gt;      [5]       0,0,0    2.04538                0.906974       1.50262</span></span>
<span><span class="co">#&gt;      ...         ...        ...                     ...           ...</span></span>
<span><span class="co">#&gt;   [1178]       0,0,0    3.64734                0.985123       2.10353</span></span>
<span><span class="co">#&gt;   [1179]       0,0,0    3.68298                0.995663       1.72712</span></span>
<span><span class="co">#&gt;   [1180]       0,0,0    2.18036                1.000000       1.83207</span></span>
<span><span class="co">#&gt;   [1181]       0,0,0    3.97304                0.983620       2.38127</span></span>
<span><span class="co">#&gt;   [1182]       0,0,0    5.15098                1.000000       3.78385</span></span>
<span><span class="co">#&gt;          Avg_diffMaxNeihgboor_1 Avg_diffMaxNeihgboor_2       avg       std</span></span>
<span><span class="co">#&gt;                       &lt;numeric&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]               0.784954               0.980075   1.85783  0.318350</span></span>
<span><span class="co">#&gt;      [2]               0.429554               0.618656   2.17273  0.273960</span></span>
<span><span class="co">#&gt;      [3]               0.828531               1.068352   1.93891  0.431294</span></span>
<span><span class="co">#&gt;      [4]               1.534711               1.594209   3.07586  0.612626</span></span>
<span><span class="co">#&gt;      [5]               0.611628               0.678974   1.83145  0.413529</span></span>
<span><span class="co">#&gt;      ...                    ...                    ...       ...       ...</span></span>
<span><span class="co">#&gt;   [1178]               1.700268               1.933103   2.69700  0.779928</span></span>
<span><span class="co">#&gt;   [1179]               2.983473               4.808131   4.94031  1.566966</span></span>
<span><span class="co">#&gt;   [1180]               0.209327               0.392278   1.80495  0.101174</span></span>
<span><span class="co">#&gt;   [1181]               1.612138               1.768132   2.67911  0.721396</span></span>
<span><span class="co">#&gt;   [1182]               0.871347               1.395824   3.64185  0.561411</span></span>
<span><span class="co">#&gt;              value</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]   2.55557</span></span>
<span><span class="co">#&gt;      [2]   2.55456</span></span>
<span><span class="co">#&gt;      [3]   2.67538</span></span>
<span><span class="co">#&gt;      [4]   4.44005</span></span>
<span><span class="co">#&gt;      [5]   2.37512</span></span>
<span><span class="co">#&gt;      ...       ...</span></span>
<span><span class="co">#&gt;   [1178]   4.20835</span></span>
<span><span class="co">#&gt;   [1179]   7.59229</span></span>
<span><span class="co">#&gt;   [1180]   1.99102</span></span>
<span><span class="co">#&gt;   [1181]   4.11212</span></span>
<span><span class="co">#&gt;   [1182]   4.41638</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3596 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p><em>Exercise</em>: How could we change this code to find transient
loops, or loops that appeared after 6 hours of PMA treatment and
dissolved after 72 hours of treatment?</p>
<p>Now let’s return to our differential looping analysis to learn how to
find quantitative changes in looping.</p>
</div>
<div class="section level3">
<h3 id="finding-differential-loops">Finding differential loops<a class="anchor" aria-label="anchor" href="#finding-differential-loops"></a>
</h3>
<p>In the previous section, we found loops that formed <em>de</em>
<em>novo</em> after treatment with PMA for 0, 6, or 72 hours. Using this
approach relies entirely on the loop caller’s threshold for calling
loops, but is unable to find quantitative changes - such as the
strengthening or weakening of existing loops between conditions.
Additionally, the presence/absence of a loop doesn’t necessarily mean
that we gained or lost a loop. In this section we take a more
statistically rigorous approach to finding differential looping.</p>
<p>To do this we will use <code>DESeq2</code> to conduct the
differential analysis. While <code>DEseq2</code> is typically used for
RNA-seq analysis, we find that it works very well for finding
differential Hi-C interactions. <code>DEseq2</code> requires three
pieces of information to run: 1) a count matrix where each row
represents a loop and each column represents a Hi-C replicate sample, 2)
a <code>data.frame</code> called <code>colData</code> that contains
sample-specific information, and 3) a design formula that describes the
model to test.</p>
<p>We can use <code>mariner</code> to create the count matrix by
extracting unnormlized contact frequency between interactions for each
Hi-C sample. First we read in the sample sheet containing Hi-C sample
information and file paths each <code>.hic</code> file:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/samples/samplesheet.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span>, colClasses<span class="op">=</span><span class="st">"factor"</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">samples</span>, <span class="st">"DataFrame"</span><span class="op">)</span></span>
<span><span class="va">samples</span></span>
<span><span class="co">#&gt; DataFrame with 12 rows and 5 columns</span></span>
<span><span class="co">#&gt;       sample condition replicate PMAtreatment                   filepath</span></span>
<span><span class="co">#&gt;     &lt;factor&gt;  &lt;factor&gt;  &lt;factor&gt;     &lt;factor&gt;                   &lt;factor&gt;</span></span>
<span><span class="co">#&gt; 1    sample1   control         1          0hr  data/hic/0hr_1_1_chr1.hic</span></span>
<span><span class="co">#&gt; 2    sample2   control         2          0hr  data/hic/0hr_1_2_chr1.hic</span></span>
<span><span class="co">#&gt; 3    sample3   control         3          0hr  data/hic/0hr_2_1_chr1.hic</span></span>
<span><span class="co">#&gt; 4    sample4   control         4          0hr  data/hic/0hr_2_2_chr1.hic</span></span>
<span><span class="co">#&gt; 5    sample5   control         5          0hr  data/hic/0hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt; ...      ...       ...       ...          ...                        ...</span></span>
<span><span class="co">#&gt; 8   sample8  treatment         2         72hr data/hic/72hr_1_2_chr1.hic</span></span>
<span><span class="co">#&gt; 9   sample9  treatment         3         72hr data/hic/72hr_2_1_chr1.hic</span></span>
<span><span class="co">#&gt; 10  sample10 treatment         4         72hr data/hic/72hr_2_2_chr1.hic</span></span>
<span><span class="co">#&gt; 11  sample11 treatment         5         72hr data/hic/72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt; 12  sample12 treatment         6         72hr data/hic/72hr_3_2_chr1.hic</span></span></code></pre></div>
<p>Since we want a count matrix with a single value for each 5 Kb loop
pixel and <code>.hic</code> file, the <code><a href="http://ericscottdavis.com/mariner/reference/pullHicPixels.html" class="external-link">pullHicPixels()</a></code>
function is what we should use with a <code>binSize</code> of 5 Kb:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Hi-C file paths from samplesheet</span></span>
<span><span class="va">hicFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">filepath</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract pixels</span></span>
<span><span class="va">pixels</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pullHicPixels.html" class="external-link">pullHicPixels</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">mergedLoops</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  blockSize<span class="op">=</span><span class="fl">100e6</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pixels</span></span>
<span><span class="co">#&gt; class: InteractionMatrix </span></span>
<span><span class="co">#&gt; dim: count matrix with 2339 interactions and 12 file(s)</span></span>
<span><span class="co">#&gt; metadata(3): binSize norm matrix</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(9): color APScoreAvg ... std value</span></span>
<span><span class="co">#&gt; colnames(12): 0hr_1_1_chr1.hic 0hr_1_2_chr1.hic ... 72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt;   72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt; colData names(2): files fileNames</span></span>
<span><span class="co">#&gt; type: MergedGInteractions</span></span>
<span><span class="co">#&gt; regions: 3596</span></span></code></pre></div>
<p>The object that results is called an <code>InteractionMatrix</code>
because it returns a single matrix for each interaction and
<code>.hic</code> file. You can access this matrix with the
<code>counts()</code> accessor:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;2339 x 12&gt; DelayedMatrix object of type "double":</span></span>
<span><span class="co">#&gt;          0hr_1_1_chr1.hic  0hr_1_2_chr1.hic ... 72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt;    [1,]                15                 6   .                10</span></span>
<span><span class="co">#&gt;    [2,]                 9                 7   .                 6</span></span>
<span><span class="co">#&gt;    [3,]                15                 7   .                 7</span></span>
<span><span class="co">#&gt;    [4,]                11                10   .                12</span></span>
<span><span class="co">#&gt;    [5,]                17                21   .                 5</span></span>
<span><span class="co">#&gt;     ...                 .                 .   .                 .</span></span>
<span><span class="co">#&gt; [2335,]                 1                 1   .                 1</span></span>
<span><span class="co">#&gt; [2336,]                 1                 0   .                 1</span></span>
<span><span class="co">#&gt; [2337,]                15                 8   .                 9</span></span>
<span><span class="co">#&gt; [2338,]                 8                 6   .                 4</span></span>
<span><span class="co">#&gt; [2339,]                 8                10   .                 9</span></span>
<span><span class="co">#&gt;         72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt;    [1,]                 8</span></span>
<span><span class="co">#&gt;    [2,]                 5</span></span>
<span><span class="co">#&gt;    [3,]                 5</span></span>
<span><span class="co">#&gt;    [4,]                 7</span></span>
<span><span class="co">#&gt;    [5,]                 3</span></span>
<span><span class="co">#&gt;     ...                 .</span></span>
<span><span class="co">#&gt; [2335,]                 0</span></span>
<span><span class="co">#&gt; [2336,]                 6</span></span>
<span><span class="co">#&gt; [2337,]                16</span></span>
<span><span class="co">#&gt; [2338,]                10</span></span>
<span><span class="co">#&gt; [2339,]                 7</span></span></code></pre></div>
<p><img src="img/interactionMatrixFig.png" width="50%" style="display: block; margin: auto;"></p>
<p>These counts are stored on-disk in an HDF5 file to conserve R-memory
usage. But we will talk more about this later. For now, we can create a
<code>DESeqDataSet</code> and run <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>  countData<span class="op">=</span><span class="fu">counts</span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span>,</span>
<span>  colData<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span>, <span class="va">samples</span><span class="op">)</span>,</span>
<span>  design<span class="op">=</span><span class="op">~</span><span class="va">replicate</span><span class="op">+</span><span class="va">condition</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">dds</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 2339 12 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(4): counts mu H cooks</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(42): baseMean baseVar ... deviance maxCooks</span></span>
<span><span class="co">#&gt; colnames(12): 0hr_1_1_chr1.hic 0hr_1_2_chr1.hic ... 72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt;   72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt; colData names(8): files fileNames ... filepath sizeFactor</span></span></code></pre></div>
<p>Get shrunken results:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>, coef<span class="op">=</span><span class="st">"condition_treatment_vs_control"</span>, type<span class="op">=</span><span class="st">"apeglm"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; out of 2336 with nonzero total read count</span></span>
<span><span class="co">#&gt; adjusted p-value &lt; 0.1</span></span>
<span><span class="co">#&gt; LFC &gt; 0 (up)       : 24, 1%</span></span>
<span><span class="co">#&gt; LFC &lt; 0 (down)     : 11, 0.47%</span></span>
<span><span class="co">#&gt; outliers [1]       : 0, 0%</span></span>
<span><span class="co">#&gt; low counts [2]     : 861, 37%</span></span>
<span><span class="co">#&gt; (mean count &lt; 3)</span></span>
<span><span class="co">#&gt; [1] see 'cooksCutoff' argument of ?results</span></span>
<span><span class="co">#&gt; [2] see 'independentFiltering' argument of ?results</span></span></code></pre></div>
<p>And inspect the results with an MA plot:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotMA.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>or a PCA plot:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html" class="external-link">varianceStabilizingTransformation</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span>intgroup<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>We can add these results from <code>DESeq2</code> to our
<code>InteractionMatrix</code> object.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span></span>
<span><span class="va">pixels</span></span>
<span><span class="co">#&gt; class: InteractionMatrix </span></span>
<span><span class="co">#&gt; dim: count matrix with 2339 interactions and 12 file(s)</span></span>
<span><span class="co">#&gt; metadata(3): binSize norm matrix</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(5): baseMean log2FoldChange lfcSE pvalue padj</span></span>
<span><span class="co">#&gt; colnames(12): 0hr_1_1_chr1.hic 0hr_1_2_chr1.hic ... 72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt;   72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt; colData names(2): files fileNames</span></span>
<span><span class="co">#&gt; type: MergedGInteractions</span></span>
<span><span class="co">#&gt; regions: 3596</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="aggregating-interactions">Aggregating interactions<a class="anchor" aria-label="anchor" href="#aggregating-interactions"></a>
</h2>
<p><code>mariner</code> is not limited to extracting single pixels. The
<code><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices()</a></code> function allows you to define square or
rectangular regions and extract them as matrices divided into specified
bins. For example, the figure below shows that supplying a
<code>binSize</code> that is half of each anchor width produces a 2x2
matrix of pixels.</p>
<p><img src="img/matricesFigure.png" width="50%" style="display: block; margin: auto;"></p>
<p>In the following sections, we will define square regions around our
merged loop pixels and extract count matrices. We will then aggregate
these matrices and visualize the results in what is called an Aggregate
Peak Analysis (APA). APA plots are essentially pile-up signal plots for
Hi-C. APA is helpful for identifying genome-wide trends at specific
loci, such as loop anchors or protein binding sites.</p>
<div class="section level3">
<h3 id="aggregate-peak-analysis-of-differential-loops">Aggregate peak analysis of differential loops<a class="anchor" aria-label="anchor" href="#aggregate-peak-analysis-of-differential-loops"></a>
</h3>
<p>We can use a helper function called <code><a href="http://ericscottdavis.com/mariner/reference/pixelsToMatrices.html" class="external-link">pixelsToMatrices()</a></code>
to expand square regions around each central pixel.</p>
<p><img src="img/pixToMat.png" width="10%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pixelsToMatrices.html" class="external-link">pixelsToMatrices</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu">interactions</span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span>, buffer<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">regions</span></span>
<span><span class="co">#&gt; MergedGInteractions object with 2339 interactions and 5 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;      [1]      chr1 100250000-100305000 ---      chr1 100320000-100375000 |</span></span>
<span><span class="co">#&gt;      [2]      chr1   92830000-92885000 ---      chr1   93035000-93090000 |</span></span>
<span><span class="co">#&gt;      [3]      chr1   77950000-78005000 ---      chr1   78085000-78140000 |</span></span>
<span><span class="co">#&gt;      [4]      chr1   89475000-89530000 ---      chr1   89575000-89630000 |</span></span>
<span><span class="co">#&gt;      [5]      chr1 212980000-213035000 ---      chr1 213045000-213100000 |</span></span>
<span><span class="co">#&gt;      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">#&gt;   [2335]      chr1 148110000-148165000 ---      chr1 148225000-148280000 |</span></span>
<span><span class="co">#&gt;   [2336]      chr1 197495000-197550000 ---      chr1 198660000-198715000 |</span></span>
<span><span class="co">#&gt;   [2337]      chr1 235490000-235545000 ---      chr1 235570000-235625000 |</span></span>
<span><span class="co">#&gt;   [2338]      chr1 203015000-203070000 ---      chr1 203310000-203365000 |</span></span>
<span><span class="co">#&gt;   [2339]      chr1 172425000-172480000 ---      chr1 172695000-172750000 |</span></span>
<span><span class="co">#&gt;           baseMean log2FoldChange     lfcSE    pvalue      padj</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;      [1]   9.56156     -0.1348939  0.240144 0.3056335  0.626778</span></span>
<span><span class="co">#&gt;      [2]   6.12458     -0.0711993  0.232172 0.5406117  0.797508</span></span>
<span><span class="co">#&gt;      [3]   6.63638     -0.2013501  0.293410 0.1451872  0.485955</span></span>
<span><span class="co">#&gt;      [4]   7.12104     -0.1768913  0.273015 0.1888885  0.525269</span></span>
<span><span class="co">#&gt;      [5]   7.91831     -0.3024159  0.386043 0.0560362  0.333280</span></span>
<span><span class="co">#&gt;      ...       ...            ...       ...       ...       ...</span></span>
<span><span class="co">#&gt;   [2335]   1.10863     -0.0057729  0.256525  0.857400        NA</span></span>
<span><span class="co">#&gt;   [2336]   1.24407      0.0572758  0.265982  0.122260        NA</span></span>
<span><span class="co">#&gt;   [2337]   9.81880      0.1302846  0.236918  0.321978  0.648978</span></span>
<span><span class="co">#&gt;   [2338]   4.57586      0.0652708  0.238179  0.546254  0.799330</span></span>
<span><span class="co">#&gt;   [2339]   7.59337      0.0545398  0.222542  0.649178  0.851997</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 3596 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Then we can use the <code><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices()</a></code> function to
extract 11 x 11 matrices by setting the <code>binSize</code> to 5
Kb:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">regions</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  blockSize<span class="op">=</span><span class="fl">100e6</span></span>
<span><span class="op">)</span></span>
<span><span class="va">matrices</span></span>
<span><span class="co">#&gt; class: InteractionArray </span></span>
<span><span class="co">#&gt; dim: 2339 interaction(s), 12 file(s), 11x11 count matrix(es)</span></span>
<span><span class="co">#&gt; metadata(3): binSize norm matrix</span></span>
<span><span class="co">#&gt; assays(3): counts rownames colnames</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(5): baseMean log2FoldChange lfcSE pvalue padj</span></span>
<span><span class="co">#&gt; colnames(12): 0hr_1_1_chr1.hic 0hr_1_2_chr1.hic ... 72hr_3_1_chr1.hic</span></span>
<span><span class="co">#&gt;   72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt; colData names(2): files fileNames</span></span>
<span><span class="co">#&gt; type: MergedGInteractions</span></span>
<span><span class="co">#&gt; regions: 3596</span></span></code></pre></div>
<p>This results in an <code>InteractionArray</code> object that is very
similar to the <code>InteractionMatrix</code> object except that it
contains an array of pixel by Hi-C count matrices that are accessible
with the <code>counts()</code> accessor:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">matrices</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;11 x 11 x 2339 x 12&gt; DelayedArray object of type "double":</span></span>
<span><span class="co">#&gt; ,,1,0hr_1_1_chr1.hic</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,10] [,11]</span></span>
<span><span class="co">#&gt;  [1,]    15     9    16   .    11     8</span></span>
<span><span class="co">#&gt;  [2,]    11     4    13   .     8    10</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [10,]    28    17    31   .     7     8</span></span>
<span><span class="co">#&gt; [11,]    32    22    33   .    11     6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ,,2339,72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,10] [,11]</span></span>
<span><span class="co">#&gt;  [1,]     2     4     7   .     7     3</span></span>
<span><span class="co">#&gt;  [2,]     4     7     7   .     7     6</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [10,]     9     8     4   .     2     0</span></span>
<span><span class="co">#&gt; [11,]     4     5     5   .     3     2</span></span></code></pre></div>
<p>This array contains count matrices in the first two dimensions,
interactions in the third dimension, and Hi-C files in the fourth
dimension. The data is stored in an HDF5 file that is accessible with
the <code>DelayedArray</code> package. This allows us to use this object
as if it were held in R’s memory, but only loading data as needed
(i.e. when operations are applied to it).</p>
<p><img src="img/interactionArrayFig.png" width="75%" style="display: block; margin: auto;"></p>
<p>We can then aggregate this array with the fast and efficient
<code><a href="http://ericscottdavis.com/mariner/reference/aggHicMatrices.html" class="external-link">aggHicMatrices()</a></code> function. This allows you to aggregate by
interactions, files, or all.</p>
<p><img src="img/aggMatFigAlt.png" width="75%" style="display: block; margin: auto;"></p>
<p>Here we will combine all interactions for each Hi-C file by setting
<code>by="files"</code>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggMats</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/aggHicMatrices.html" class="external-link">aggHicMatrices</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">matrices</span>, by<span class="op">=</span><span class="st">"files"</span><span class="op">)</span></span>
<span><span class="va">aggMats</span></span>
<span><span class="co">#&gt; &lt;11 x 11 x 12&gt; DelayedArray object of type "double":</span></span>
<span><span class="co">#&gt; ,,0hr_1_1_chr1.hic</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,10] [,11]</span></span>
<span><span class="co">#&gt;  [1,]  8016  8388  8627   .  7338  6923</span></span>
<span><span class="co">#&gt;  [2,]  8850  9191  9261   .  7859  7061</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [10,] 20223 18791 18770   .  9213  8374</span></span>
<span><span class="co">#&gt; [11,] 24015 20171 18479   .  9130  8225</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ,,72hr_3_2_chr1.hic</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,10] [,11]</span></span>
<span><span class="co">#&gt;  [1,]  7084  7219  7451   .  6353  6019</span></span>
<span><span class="co">#&gt;  [2,]  7679  7994  8120   .  6767  6299</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [10,] 17511 16543 17877   .  8062  7226</span></span>
<span><span class="co">#&gt; [11,] 22929 17669 16094   .  7939  6986</span></span></code></pre></div>
<p><em>Exercise</em> - Try different values for the <code>by</code>
argument and see what you get!</p>
<p>Since the result is a <code>DelayedArray</code> object, we can use
<code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code> functions to do more complex aggregations. Let’s
combine all the biological and technical replicates for each condition
(0hr or 72hr).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">aggMats</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">aggMats</span><span class="op">[</span>,,<span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>And visualize them with the <code><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix()</a></code> function:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">ctr</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">trt</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-29-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-29-2.png" width="50%"></p>
<p>You’ll notice that there is an artifact in the bottom left-hand
corner. This is because interactions that are short, or close to the
diagonal, are being superimposed. The <code><a href="http://ericscottdavis.com/mariner/reference/removeShortPairs.html" class="external-link">removeShortPairs()</a></code>
function can be used upstream of <code><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices()</a></code> function
to remove any part of the regions that would cross the diagonal. The
following code shows how these functions can be chained to produce APA
plots without artifacts caused by the Hi-C diagonal.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Remove short pairs &amp; extract counts</span></span>
<span><span class="va">matrices</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">regions</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/removeShortPairs.html" class="external-link">removeShortPairs</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices</a></span><span class="op">(</span>files<span class="op">=</span><span class="va">hicFiles</span>, binSize<span class="op">=</span><span class="fl">5e3</span>, blockSize<span class="op">=</span><span class="fl">100e6</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Aggregate by files</span></span>
<span><span class="va">aggMats</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/aggHicMatrices.html" class="external-link">aggHicMatrices</a></span><span class="op">(</span><span class="va">matrices</span>, by<span class="op">=</span><span class="st">"files"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Combine bio &amp; tech replicates</span></span>
<span><span class="va">ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">aggMats</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">aggMats</span><span class="op">[</span>,,<span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize without artifacts</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">ctr</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">trt</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-30-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-30-2.png" width="50%"></p>
<p>You may notice that there is not much difference between the APA of
the control vs. treatment because we are aggregating all loops together.
We can subset the <code>InteractionArray</code> object according to our
differential looping results to show genome-wide differences that
support our analysis.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset into gained/lost loops</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">matrices</span><span class="op">)</span></span>
<span><span class="va">gained</span> <span class="op">&lt;-</span> <span class="va">matrices</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">lost</span> <span class="op">&lt;-</span> <span class="va">matrices</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Aggregate gained/lost for control &amp; treated</span></span>
<span><span class="va">gained_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">gained</span><span class="op">)</span><span class="op">[</span>,,,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">gained_trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">gained</span><span class="op">)</span><span class="op">[</span>,,,<span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">lost_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">lost</span><span class="op">)</span><span class="op">[</span>,,,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">lost_trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">lost</span><span class="op">)</span><span class="op">[</span>,,,<span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="va">sum</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Normalize to the number of interactions</span></span>
<span><span class="va">gained_ctr</span> <span class="op">&lt;-</span> <span class="va">gained_ctr</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gained</span><span class="op">)</span></span>
<span><span class="va">gained_trt</span> <span class="op">&lt;-</span> <span class="va">gained_trt</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gained</span><span class="op">)</span></span>
<span><span class="va">lost_ctr</span> <span class="op">&lt;-</span> <span class="va">lost_ctr</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lost</span><span class="op">)</span></span>
<span><span class="va">lost_trt</span> <span class="op">&lt;-</span> <span class="va">lost_trt</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lost</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Find common scale for gained/lost</span></span>
<span><span class="va">gained_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">gained_ctr</span>, <span class="va">gained_trt</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lost_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">lost_ctr</span>, <span class="va">lost_trt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then we can visualize the gained loops in control and treatment:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_ctr</span>, zrange<span class="op">=</span><span class="va">gained_scale</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_trt</span>, zrange<span class="op">=</span><span class="va">gained_scale</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-32-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-32-2.png" width="50%"></p>
<p>And the lost loops in control and treatment:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">lost_ctr</span>, zrange<span class="op">=</span><span class="va">lost_scale</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">lost_trt</span>, zrange<span class="op">=</span><span class="va">lost_scale</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-33-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-33-2.png" width="50%"></p>
<p>In the <a href="vis_with_pg">final section of this workshop</a> we
will take these plots and visualize them on a single page with
labels.</p>
<p><code>mariner</code> includes the <code><a href="http://ericscottdavis.com/mariner/reference/pileupPixels.html" class="external-link">pileupPixels()</a></code>
function, which wraps several of these steps together to simply your
workflow. For example, the code below can be used to accomplish the same
task.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Identify gained/lost loops</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pixels</span><span class="op">)</span></span>
<span><span class="va">gained</span> <span class="op">&lt;-</span> <span class="va">pixels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">lost</span> <span class="op">&lt;-</span> <span class="va">pixels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Calculate apa matrices</span></span>
<span><span class="va">gained_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupPixels.html" class="external-link">pileupPixels</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu">interactions</span><span class="op">(</span><span class="va">gained</span><span class="op">)</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gained_trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupPixels.html" class="external-link">pileupPixels</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu">interactions</span><span class="op">(</span><span class="va">gained</span><span class="op">)</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  minPairDist<span class="op">=</span><span class="fl">50e3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lost_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupPixels.html" class="external-link">pileupPixels</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu">interactions</span><span class="op">(</span><span class="va">lost</span><span class="op">)</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lost_trt</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupPixels.html" class="external-link">pileupPixels</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="fu">interactions</span><span class="op">(</span><span class="va">lost</span><span class="op">)</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Find common scale for gained/lost</span></span>
<span><span class="va">gained_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">gained_ctr</span>, <span class="va">gained_trt</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lost_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">lost_ctr</span>, <span class="va">lost_trt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then we can visualize the gained loops in control and treatment:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_ctr</span>, zrange<span class="op">=</span><span class="va">gained_scale</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_trt</span>, zrange<span class="op">=</span><span class="va">gained_scale</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-35-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-35-2.png" width="50%"></p>
<p>And the lost loops in control and treatment:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">lost_ctr</span>, zrange<span class="op">=</span><span class="va">lost_scale</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">lost_trt</span>, zrange<span class="op">=</span><span class="va">lost_scale</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-36-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-36-2.png" width="50%"></p>
<p>Running each step separately allows you to have the maximum amount of
control over the settings which improves efficiency and performance. It
also gives you more flexibility to allow custom functions that haven’t
been explicitly created.</p>
</div>
<div class="section level3">
<h3 id="aggregate-domain-analysis-of-differential-loops">Aggregate domain analysis of differential loops<a class="anchor" aria-label="anchor" href="#aggregate-domain-analysis-of-differential-loops"></a>
</h3>
<p>In addition to pileups of pixels, we can also perform pileup analysis
for domains. In this analysis, we will aggregate the whole domain
encompassed by the differential loops so we can visualize how the
domains are changing.</p>
<p>We can build a helper function that makes TAD with a buffer on either
side from the loops.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Write a helper function to convert</span></span>
<span><span class="co">## loops into TADs</span></span>
<span><span class="va">makeTadsFromLoops</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">InteractionSet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/distances.html" class="external-link">pairdist</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">'span'</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/GInteractions-accessors.html" class="external-link">seqnames1</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/GInteractions-accessors.html" class="external-link">start1</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">-</span><span class="va">d</span>,</span>
<span>    <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/GInteractions-accessors.html" class="external-link">end2</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">+</span><span class="va">d</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/as_ginteractions.html" class="external-link">as_ginteractions</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">tads</span> <span class="op">&lt;-</span> <span class="fu">makeTadsFromLoops</span><span class="op">(</span><span class="va">mergedLoops</span><span class="op">)</span></span></code></pre></div>
<p>Then we can filter these tads for those that are gained after 72
hours of PMA treatment.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Identify gained TADs</span></span>
<span><span class="va">gained</span> <span class="op">&lt;-</span> <span class="va">tads</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>The <code><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices()</a></code> function can be used to extract
count matrices for these tads at a given resolution (in this case 5
Kb).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Pull TAD regions</span></span>
<span><span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pullHicMatrices.html" class="external-link">pullHicMatrices</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">gained</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  blockSize<span class="op">=</span><span class="fl">100e6</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Pairs are not binned to `binSize`.</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Snapping to binSize=5000, </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">ℹ</span> Use `assignToBins()` or `snapToBins()` for more binning options.</span></span></code></pre></div>
<p>Because each region is a different size, an irregular or jagged array
object is returned:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mats</span></span>
<span><span class="co">#&gt; class: InteractionJaggedArray</span></span>
<span><span class="co">#&gt; dim: 24 interaction(s), 6 file(s), variable count matrix(es)</span></span>
<span><span class="co">#&gt; metadata(3): binSize, norm, matrix</span></span>
<span><span class="co">#&gt; colData: </span></span>
<span><span class="co">#&gt; colData names(2): files, fileNames</span></span>
<span><span class="co">#&gt; HDF5: /tmp/RtmpFKsiBS/file4144874e33e.h5</span></span></code></pre></div>
<p>The <code>counts()</code> function gives access to these count
matrices, which can be realized with <code><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list()</a></code> if
needed.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">mats</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;n x m x 24 x 6&gt; JaggedArray:</span></span>
<span><span class="co">#&gt; ,,1,1</span></span>
<span><span class="co">#&gt; &lt;55 x 55&gt; matrix</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,54] [,55]</span></span>
<span><span class="co">#&gt;  [1,]   108    35    18   .     1     1</span></span>
<span><span class="co">#&gt;  [2,]    35   112    29   .     2     1</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [54,]     1     2     0   .   108    34</span></span>
<span><span class="co">#&gt; [55,]     1     1     1   .    34   109</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ,,24,6</span></span>
<span><span class="co">#&gt; &lt;45 x 45&gt; matrix</span></span>
<span><span class="co">#&gt;        [,1]  [,2]  [,3] ... [,44] [,45]</span></span>
<span><span class="co">#&gt;  [1,]   176    32    14   .     1     4</span></span>
<span><span class="co">#&gt;  [2,]    32   213    32   .     1     3</span></span>
<span><span class="co">#&gt;   ...     .     .     .   .     .     .</span></span>
<span><span class="co">#&gt; [44,]     1     1     0   .   179    30</span></span>
<span><span class="co">#&gt; [45,]     4     3     1   .    30   251</span></span></code></pre></div>
<p>For an aggregate TAD analysis, we need to make these irregular arrays
the same size. The <code><a href="http://ericscottdavis.com/mariner/reference/regularize.html" class="external-link">regularize()</a></code> function resizes each
matrix to the rows and column dimensions specified in the
<code>ndim</code> parameter. The <code>scale</code> parameter will scale
each matrix to its total signal so that each matrix contributes the same
relative signal to the aggregation. The result is an
<code>InteractionArray</code> as we’ve seen in previous sections.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Convert to regular array</span></span>
<span><span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/regularize.html" class="external-link">regularize</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">mats</span>,</span>
<span>  ndim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  nBlocks<span class="op">=</span><span class="fl">10</span>,</span>
<span>  scale<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The regularized array can then be aggregated and visualized like
before:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Aggregate</span></span>
<span><span class="va">agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/aggHicMatrices.html" class="external-link">aggHicMatrices</a></span><span class="op">(</span><span class="va">reg</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">agg</span>, zrange<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-43-1.png" width="700"></p>
<p>We can use the wrapper function <code><a href="http://ericscottdavis.com/mariner/reference/pileupDomains.html" class="external-link">pileupDomains()</a></code> to carry
out these steps for us:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Identify gained TADs</span></span>
<span><span class="va">gained</span> <span class="op">&lt;-</span> <span class="va">tads</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Calculate aggregate TAD plots</span></span>
<span><span class="va">gained_ctr_tads</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupDomains.html" class="external-link">pileupDomains</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">gained</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  buffer<span class="op">=</span><span class="fl">0.25</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gained_trt_tads</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/pileupDomains.html" class="external-link">pileupDomains</a></span><span class="op">(</span></span>
<span>  x<span class="op">=</span><span class="va">gained</span>,</span>
<span>  files<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span>,</span>
<span>  binSize<span class="op">=</span><span class="fl">5e3</span>,</span>
<span>  buffer<span class="op">=</span><span class="fl">0.25</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Find common scale for gained/lost</span></span>
<span><span class="va">gq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">gained_ctr_tads</span>, <span class="va">gained_trt_tads</span><span class="op">)</span>, <span class="fl">0.90</span><span class="op">)</span></span>
<span><span class="va">gained_tad_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">gq</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_ctr_tads</span>, zrange<span class="op">=</span><span class="va">gained_tad_scale</span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span><span class="va">gained_trt_tads</span>, zrange<span class="op">=</span><span class="va">gained_tad_scale</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-44-1.png" width="50%"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-44-2.png" width="50%"></p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-with-plotgardener">Visualizing with <code>plotgardener</code>
<a id="vis_with_pg"></a><a class="anchor" aria-label="anchor" href="#visualizing-with-plotgardener"></a>
</h2>
<p><code>plotgardener</code> is a genomic data visualization package for
R. It uses a coordinate-based plotting system to give users precise
control over the size and placement of visualizations. In this section
we will use <code>plotgardener</code> to programmatically place our APA
plots on a page and add text labels.</p>
<p>To visualize multiple plots with <code>plotgardener</code> you start
by creating a page:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener" class="external-link">plotgardener</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create page</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html" class="external-link">pageCreate</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">3</span>, height<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p>We can pass additional parameters to the <code><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix()</a></code>
function specifying where we would like these to be placed on the
page.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## APA plots</span></span>
<span><span class="va">gainedApa</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">gained_ctr</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">gained_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">gained_trt</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">gained_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="va">lostApa</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">lost_ctr</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">lost_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">lost_trt</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">lost_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>We can use annotation functions to add scale legends.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## APA heatmap legends</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoHeatmapLegend.html" class="external-link">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">gainedApa</span>,</span>
<span>  orientation<span class="op">=</span><span class="st">'h'</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoHeatmapLegend.html" class="external-link">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">lostApa</span>,</span>
<span>  orientation<span class="op">=</span><span class="st">'h'</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<p>and text labels:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot text labels</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"72hr-loops"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.975</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"0hr-loops"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.975</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"0hr-hic"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.975</span>,</span>
<span>  rot<span class="op">=</span><span class="fl">90</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"72hr-hic"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.975</span>,</span>
<span>  rot<span class="op">=</span><span class="fl">90</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>When the plot is to your liking, you can remove the guides on the
page:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageGuideHide.html" class="external-link">pageGuideHide</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-50-1.png" width="700"><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-50-2.png" width="700"></p>
<p><code>plotgardener</code> is a very useful tool for making
publication-quality genomic figures. Visit the <a href="https://phanstiellab.github.io/plotgardener/" class="external-link">website</a> to learn
more about what you can do with <code>plotgardener</code>.</p>
<p>For example, we could plot an example loop beside these APA
plots:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create page</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html" class="external-link">pageCreate</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">6</span>, height<span class="op">=</span><span class="fl">3</span>, showGuides<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## APA plots</span></span>
<span><span class="va">gainedApa</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">gained_ctr</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">gained_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">gained_trt</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">gained_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="va">lostApa</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">lost_ctr</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">lost_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="http://ericscottdavis.com/mariner/reference/plotMatrix.html" class="external-link">plotMatrix</a></span><span class="op">(</span></span>
<span>  data<span class="op">=</span><span class="va">lost_trt</span>,</span>
<span>  zrange<span class="op">=</span><span class="va">lost_scale</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.95</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## APA heatmap legends</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoHeatmapLegend.html" class="external-link">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">gainedApa</span>,</span>
<span>  orientation<span class="op">=</span><span class="st">'h'</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoHeatmapLegend.html" class="external-link">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">lostApa</span>,</span>
<span>  orientation<span class="op">=</span><span class="st">'h'</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>  height<span class="op">=</span><span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot text labels</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"72hr-loops"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.975</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"0hr-loops"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">1.975</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"0hr-hic"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.975</span>,</span>
<span>  rot<span class="op">=</span><span class="fl">90</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"72hr-hic"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  y<span class="op">=</span><span class="fl">1.975</span>,</span>
<span>  rot<span class="op">=</span><span class="fl">90</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot Hi-C example loop</span></span>
<span><span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotHicSquare.html" class="external-link">plotHicSquare</a></span><span class="op">(</span></span>
<span>  <span class="co">## Path to .hic file</span></span>
<span>  data<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  </span>
<span>  <span class="co">## Loci</span></span>
<span>  chrom<span class="op">=</span><span class="st">"chr1"</span>,</span>
<span>  chromstart<span class="op">=</span><span class="fl">213.9e6</span>,</span>
<span>  chromend<span class="op">=</span><span class="fl">215e6</span>,</span>
<span>  </span>
<span>  <span class="co">## Hi-C params</span></span>
<span>  resolution<span class="op">=</span><span class="fl">10e3</span>,</span>
<span>  zrange<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">40</span><span class="op">)</span>,</span>
<span>  norm<span class="op">=</span><span class="st">"KR"</span>,</span>
<span>  half<span class="op">=</span><span class="st">"top"</span>,</span>
<span>  </span>
<span>  <span class="co">## Placement on page</span></span>
<span>  x<span class="op">=</span><span class="fl">3</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">2</span>,</span>
<span>  height<span class="op">=</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotHicSquare.html" class="external-link">plotHicSquare</a></span><span class="op">(</span></span>
<span>  <span class="co">## Path to .hic file</span></span>
<span>  data<span class="op">=</span><span class="va">hicFiles</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span>,</span>
<span>  </span>
<span>  <span class="co">## Loci</span></span>
<span>  chrom<span class="op">=</span><span class="st">"chr1"</span>,</span>
<span>  chromstart<span class="op">=</span><span class="fl">213.9e6</span>,</span>
<span>  chromend<span class="op">=</span><span class="fl">215e6</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting params</span></span>
<span>  resolution<span class="op">=</span><span class="fl">10e3</span>,</span>
<span>  zrange<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">40</span><span class="op">)</span>,</span>
<span>  norm<span class="op">=</span><span class="st">"KR"</span>,</span>
<span>  half<span class="op">=</span><span class="st">"bottom"</span>,</span>
<span>  </span>
<span>  <span class="co">## Placement on page</span></span>
<span>  x<span class="op">=</span><span class="fl">3</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">2</span>,</span>
<span>  height<span class="op">=</span><span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Genome coordinate label</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoGenomeLabel.html" class="external-link">annoGenomeLabel</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">hicPlot</span>,</span>
<span>  scale<span class="op">=</span><span class="st">"Mb"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">3</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Hi-C heatmap legend</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoHeatmapLegend.html" class="external-link">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>  plot<span class="op">=</span><span class="va">hicPlot</span>,</span>
<span>  orientation<span class="op">=</span><span class="st">'v'</span>,</span>
<span>  x<span class="op">=</span><span class="fl">5.05</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>  width<span class="op">=</span><span class="fl">0.1</span>,</span>
<span>  height<span class="op">=</span><span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot title &amp; half labels</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"Example Loop; 10Kb bins"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">4</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"0hrs"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">3.05</span>,</span>
<span>  y<span class="op">=</span><span class="fl">0.55</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html" class="external-link">plotText</a></span><span class="op">(</span></span>
<span>  label<span class="op">=</span><span class="st">"72hrs"</span>,</span>
<span>  x<span class="op">=</span><span class="fl">4.95</span>,</span>
<span>  y<span class="op">=</span><span class="fl">2.45</span>,</span>
<span>  just<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"right"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="Bioc2023mariner_files/figure-html/unnamed-chunk-51-2.png" width="700"></p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bond2022.10.31.514600" class="csl-entry">
Bond, Marielle L., Eric S. Davis, Ivana Y. Quiroga, Michael I. Love,
Hyejung Won, and Douglas H. Phanstiel. 2022. <span>“Chromatin Loop
Dynamics During Cellular Differentiation Are Associated with Changes to
Both Anchor and Internal Regulatory Features.”</span> <em>bioRxiv</em>.
<a href="https://doi.org/10.1101/2022.10.31.514600" class="external-link">https://doi.org/10.1101/2022.10.31.514600</a>.
</div>
<div id="ref-eric_davis_2022_7514346" class="csl-entry">
Davis, Eric, and Kathleen S Metz Reed. 2022. <em>EricSDavis/dietJuicer:
dietJuicer 1.0.1</em> (version 1.0.1). Zenodo. <a href="https://doi.org/10.5281/zenodo.7514346" class="external-link">https://doi.org/10.5281/zenodo.7514346</a>.
</div>
<div id="ref-durand_juicer_2016" class="csl-entry">
Durand, Neva C., Muhammad S. Shamim, Ido Machol, Suhas S. P. Rao, Miriam
H. Huntley, Eric S. Lander, and Erez Lieberman Aiden. 2016.
<span>“Juicer <span>Provides</span> a
<span>One</span>-<span>Click</span> <span>System</span> for
<span>Analyzing</span> <span>Loop</span>-<span>Resolution</span>
<span>Hi</span>-<span>C</span> <span>Experiments</span>.”</span>
<em>Cell Systems</em> 3 (1): 95–98. <a href="https://doi.org/10.1016/j.cels.2016.07.002" class="external-link">https://doi.org/10.1016/j.cels.2016.07.002</a>.
</div>
<div id="ref-plotgardener" class="csl-entry">
Kramer, Nicole E., Eric S. Davis, Craig D. Wenger, Erika M. Deoudes,
Sarah M. Parker, Michael I. Love, and Douglas H. Phanstiel. 2022.
<span>“Plotgardener: Cultivating Precise Multi-Panel Figures in
r.”</span> <em>Bioinformatics</em>. <a href="https://doi.org/10.1093/bioinformatics/btac057" class="external-link">https://doi.org/10.1093/bioinformatics/btac057</a>.
</div>
<div id="ref-DESeq2" class="csl-entry">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014.
<span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq
Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-InteractionSet" class="csl-entry">
Lun, Aaron T L, Malcolm Perry, and Elizabeth Ing-Simmons. 2016.
<span>“Infrastructure for Genomic Interactions: Bioconductor Classes for
Hi-c, ChIA-PET and Related Experiments.”</span> <em>F1000Res.</em> 5:
950.
</div>
<div id="ref-DelayedArray" class="csl-entry">
Pagès, Hervé. 2023. <em>DelayedArray: A Unified Framework for Working
Transparently with on-Disk and in-Memory Array-Like Datasets</em>. <a href="https://doi.org/10.18129/B9.bioc.DelayedArray" class="external-link">https://doi.org/10.18129/B9.bioc.DelayedArray</a>.
</div>
<div id="ref-rao_3d_2014" class="csl-entry">
Rao, Suhas S.P., Miriam H. Huntley, Neva C. Durand, Elena K. Stamenova,
Ivan D. Bochkov, James T. Robinson, Adrian L. Sanborn, et al. 2014.
<span>“A <span>3D</span> <span>Map</span> of the <span>Human</span>
<span>Genome</span> at <span>Kilobase</span> <span>Resolution</span>
<span>Reveals</span> <span>Principles</span> of <span>Chromatin</span>
<span>Looping</span>.”</span> <em>Cell</em> 159 (7): 1665–80. <a href="https://doi.org/10.1016/j.cell.2014.11.021" class="external-link">https://doi.org/10.1016/j.cell.2014.11.021</a>.
</div>
<div id="ref-Rowley01032020" class="csl-entry">
Rowley, M. Jordan, Axel Poulet, Michael H. Nichols, Brianna J. Bixler,
Adrian L. Sanborn, Elizabeth A. Brouhard, Karen Hermetz, et al. 2020.
<span>“Analysis of Hi-c Data Using SIP Effectively Identifies Loops in
Organisms from c. Elegans to Mammals.”</span> <em>Genome Research</em>
30 (3): 447–58. <a href="https://doi.org/10.1101/gr.257832.119" class="external-link">https://doi.org/10.1101/gr.257832.119</a>.
</div>
<div id="ref-apeglm" class="csl-entry">
Zhu, Anqi, Joseph G. Ibrahim, and Michael I. Love. 2018.
<span>“Heavy-Tailed Prior Distributions for Sequence Count Data:
Removing the Noise and Preserving Large Differences.”</span>
<em>Bioinformatics</em>. <a href="https://doi.org/10.1093/bioinformatics/bty895" class="external-link">https://doi.org/10.1093/bioinformatics/bty895</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Eric Davis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
